Search.setIndex({"docnames": ["api/modules", "api/thoraxe", "api/thoraxe.add_transcripts", "api/thoraxe.subexons", "api/thoraxe.transcript_info", "api/thoraxe.transcript_query", "api/thoraxe.utils", "index", "installation", "introduction", "output", "programs/add_transcripts", "programs/programs", "programs/thoraxe", "programs/transcript_query"], "filenames": ["api/modules.rst", "api/thoraxe.rst", "api/thoraxe.add_transcripts.rst", "api/thoraxe.subexons.rst", "api/thoraxe.transcript_info.rst", "api/thoraxe.transcript_query.rst", "api/thoraxe.utils.rst", "index.rst", "installation.rst", "introduction.rst", "output.rst", "programs/add_transcripts.rst", "programs/programs.rst", "programs/thoraxe.rst", "programs/transcript_query.rst"], "titles": ["thoraxe", "thoraxe package", "thoraxe.add_transcripts package", "thoraxe.subexons package", "thoraxe.transcript_info package", "thoraxe.transcript_query package", "thoraxe.utils package", "ThorAxe\u2019s documentation!", "Installation", "Quick Introduction", "Output description", "add_transcripts", "thoraxe", "thoraxe", "transcript_query"], "terms": {"packag": [0, 7], "subpackag": [0, 7], "add_transcript": [0, 1, 7, 9, 12], "submodul": [0, 7], "modul": [0, 7, 8], "content": [0, 7], "subexon": [0, 1, 4, 7, 10], "align": [0, 1, 4, 8, 10, 13], "ases": [0, 1], "graph": [0, 1, 7, 9, 13], "phylosof": [0, 1, 7, 8, 13], "plot": [0, 1, 10, 13], "rescu": [0, 1, 13], "tidi": [0, 1, 10], "transcript_info": [0, 1, 3, 7], "cluster": [0, 1, 3, 13], "exon_clust": [0, 1], "phase": [0, 1, 3, 11, 13], "transcript_queri": [0, 1, 3, 7, 9, 11, 12, 13], "util": [0, 1, 3, 7], "folder": [0, 1, 3, 5, 8, 10, 13], "speci": [0, 1, 3, 4, 5, 10, 11, 13, 14], "split": [0, 1, 4, 10], "window": [0, 1, 3, 8], "add_s_exon_phases_and_coordin": [0, 1, 7], "create_chimeric_msa": [0, 1, 7], "get_s_exon_msa": [0, 1, 7], "get_s_exon": [0, 1, 7], "get_subexon": [0, 1, 7], "get_transcript": [0, 1, 7], "main": [0, 1, 2, 5, 7], "merge_clust": [0, 1, 4, 7], "parse_command_lin": [0, 1, 2, 5, 7], "update_subexon_t": [0, 1, 7], "version": [0, 5, 7, 8, 10, 11, 13, 14], "path": [1, 2, 3, 5, 6, 7, 8, 11, 13], "exont": [1, 2], "input": [1, 2, 3, 4, 7, 10, 12, 13], "sequenc": [1, 2, 3, 4, 5, 6, 8, 10, 11], "tsl": [1, 2, 4, 5, 10, 13], "add_sequ": [1, 2], "add_to_exont": [1, 2], "add_to_tsl": [1, 2], "check_input": [1, 2], "get_output_path": [1, 2], "read_fasta": [1, 2], "colclust": [1, 3], "colpattern": [1, 3], "end": [1, 3, 4, 11, 14], "pattern": [1, 3], "start": [1, 3, 4, 11], "problematicsubexonblock": [1, 3], "block_typ": [1, 3], "gap_block_end": [1, 3], "gap_block_start": [1, 3], "sequence_index": [1, 3], "subexon_block_end": [1, 3], "subexon_block_start": [1, 3], "subexon_block": [1, 3], "cluster_subexon_block": [1, 3], "column_clust": [1, 3], "column_pattern": [1, 3], "create_chimeric_sequ": [1, 3], "create_msa_matrix": [1, 3], "create_subexon_matrix": [1, 3], "delete_pad": [1, 3], "delete_subexon": [1, 3], "disintegr": [1, 3, 13], "gene2speci": [1, 3], "get_consensu": [1, 3], "get_gene_id": [1, 3], "get_subexon_boundari": [1, 3], "get_submsa": [1, 3], "impute_missing_s_exon": [1, 3], "move_block": [1, 3], "move_problematic_block_clust": [1, 3], "move_subexon_block": [1, 3], "msa2sequ": [1, 3], "msa_matric": [1, 3], "problematic_block_clust": [1, 3], "problematic_subexon_block": [1, 3], "read_msa_fasta": [1, 3], "resume_seq": [1, 3], "run_align": [1, 3], "save_s_exon": [1, 3], "score_solut": [1, 3], "sort_speci": [1, 3], "subexon_connect": [1, 3], "pathdata": [1, 3], "gene": [1, 3, 4, 5, 9, 10, 11, 13, 14], "transcript": [1, 2, 3, 4, 5, 7, 10, 11, 13, 14], "conserved_as": [1, 3], "create_ases_t": [1, 3], "detect_as": [1, 3], "get_transcript_scor": [1, 3], "read_splice_graph": [1, 3], "nodes_and_edges2genes_and_transcript": [1, 3], "splice_graph_gml": [1, 3], "get_s_exon2char": [1, 3], "get_transcript2phylosof": [1, 3], "phylosofs_input": [1, 3], "prune_tre": [1, 3], "create_python_structur": [1, 3], "plot_msa_subexon": [1, 3], "get_unaligned_region": [1, 3], "modify_subexon_clust": [1, 3], "subexon_rescue_phas": [1, 3], "interv": [1, 3], "compon": [1, 3], "create_subexon_t": [1, 3], "disjoint_interv": [1, 3], "update_to_merge_list": [1, 3], "get_tidy_t": [1, 3], "cluster2str": [1, 4], "fill_clust": [1, 4], "inform_about_delet": [1, 4], "set_to_delet": [1, 4], "coverag": [1, 4, 13], "coverage_shortest": [1, 4], "percent_ident": [1, 4], "bases_to_complete_next_codon": [1, 4], "bases_to_complete_previous_codon": [1, 4], "calculate_phas": [1, 4], "check_order_and_phas": [1, 4], "end_phase_previous_exon": [1, 4], "add_exon_sequ": [1, 4], "add_protein_seq": [1, 4], "delete_badquality_sequ": [1, 4], "delete_identical_sequ": [1, 4], "delete_incomplete_sequ": [1, 4], "find_identical_exon": [1, 4], "find_identical_sequ": [1, 4], "merge_identical_exon": [1, 4], "read_exon_fil": [1, 4], "read_transcript_info": [1, 3, 4], "read_tsl_fil": [1, 4], "store_clust": [1, 4], "dictseq2fasta": [1, 5], "filter_ortho": [1, 5], "generic_ensembl_rest_request": [1, 5], "get_biomart_exons_annot": [1, 5], "get_exons_sequ": [1, 5], "get_geneids_from_symbol": [1, 5], "get_genetre": [1, 5], "get_listofexon": [1, 5], "get_listoftranscript": [1, 5], "get_ortholog": [1, 5], "get_transcripts_ortholog": [1, 5], "is_esemble_id": [1, 5], "lodict2csv": [1, 5], "save_ensembl_vers": [1, 5], "write_tsl_fil": [1, 5], "create_subfold": [1, 6], "check_species_nam": [1, 6], "get_species_list": [1, 6], "split_exon": [1, 6], "split_length": [1, 6], "split_seq": [1, 6], "is_window": [1, 6], "pipelin": [1, 4, 5, 10], "script": [1, 2, 5, 11], "function": [1, 2, 3, 4, 5, 6, 10], "tbl": 1, "add": [1, 2, 3, 4, 7, 11], "": [1, 2, 3, 4, 6, 9, 11, 13, 14], "exon": [1, 2, 3, 4, 5, 6, 7, 9, 11, 13, 14], "genom": [1, 3, 4, 7, 10, 11], "coordin": [1, 3, 4, 11], "tabl": [1, 2, 3, 4, 11], "For": [1, 3, 4, 8, 9, 10, 11], "each": [1, 3, 4, 5, 10, 11], "s_exon_genomic_sequ": 1, "1": [1, 2, 3, 4, 5, 8, 10, 11, 13, 14], "defin": [1, 2, 3, 11], "close": 1, "s_exon_codingstart": 1, "s_exon_codingend": 1, "where": [1, 3, 4, 10], "i": [1, 2, 3, 4, 5, 6, 7, 8, 10, 11, 13, 14], "greater": 1, "equal": 1, "revers": [1, 11], "strand": [1, 11], "It": [1, 2, 3, 4, 5, 6, 8, 9, 10, 11, 13, 14], "also": [1, 3, 4, 10], "s_exon_startphas": 1, "s_exon_endphas": 1, "those": 1, "an": [1, 2, 3, 4, 5, 6, 8, 9, 10, 11, 13], "protein": [1, 3, 4, 10], "level": [1, 3, 4, 13], "can": [1, 3, 4, 7, 8, 9, 10, 11, 14], "entir": 1, "sub": [1, 3, 13], "part": 1, "In": [1, 3, 4, 8, 10], "case": [1, 10], "ident": [1, 4, 10, 13], "ar": [1, 3, 4, 5, 8, 9, 10, 13, 14], "same": [1, 3, 4, 10, 13], "howev": [1, 3, 14], "when": [1, 4, 5, 10, 13], "sever": 1, "need": [1, 3, 5, 8, 9, 10, 14], "calcul": [1, 4, 10], "exampl": [1, 5, 8, 9, 10, 11], "posit": [1, 3, 7, 12], "The": [1, 3, 4, 5, 6, 8, 9, 10, 11, 13], "follow": [1, 3, 8, 10, 11, 13], "share": [1, 7], "codon": [1, 4, 11], "previou": [1, 4, 10], "thi": [1, 3, 4, 5, 8, 9, 10, 11], "repres": [1, 10], "intron": [1, 4, 11], "lower": [1, 3], "charact": [1, 3, 4, 10], "nucleotid": [1, 11], "2": [1, 2, 3, 4, 6, 13], "respect": [1, 4, 10], "v": [1, 11, 14], "vvwwwxxxyyyzz": 1, "z": [1, 4], "becaus": [1, 3, 4], "we": [1, 4, 8, 10], "residu": [1, 3, 13], "come": [1, 9], "from": [1, 2, 3, 4, 5, 7, 10, 11, 13, 14], "vvv": 1, "assign": 1, "wai": [1, 8], "zzz": [1, 4], "begin": [1, 4], "next": [1, 4], "vwxy": 1, "let": [1, 4, 9, 11], "sai": [1, 9], "compos": [1, 5], "two": [1, 3, 4, 10, 11, 13], "vw": 1, "xy": 1, "Then": [1, 3, 8], "correspond": [1, 5, 10], "vvwww": 1, "xxxyyyzz": 1, "0": [1, 2, 3, 4, 13], "first": [1, 3, 4, 10, 11], "ha": [1, 3, 4, 6, 10, 11, 13, 14], "11111111": 1, "12222": 1, "2345": 1, "67890": 1, "12345678": 1, "90123": 1, "4": [1, 3, 4, 6, 13], "6": [1, 3, 4, 6, 8], "10": [1, 3, 4, 7, 8, 10, 13], "second": [1, 3, 4, 10, 11], "11": 1, "18": [1, 8], "amino": [1, 3], "acid": [1, 3], "5": [1, 3, 4, 13], "instead": [1, 8, 9, 11], "base": [1, 4, 11], "between": [1, 3, 4, 9, 10, 11], "particular": [1, 3, 10], "complet": [1, 4, 11], "neg": 1, "2222": 1, "11111": 1, "3210": 1, "98765": 1, "43210987": 1, "65432": 1, "If": [1, 3, 4, 6, 7, 8, 9, 10, 11, 13, 14], "belong": [1, 4], "go": [1, 4, 8, 10], "contain": [1, 2, 4, 10, 11, 14], "wxyz": 1, "therefor": [1, 4], "possibl": [1, 10, 13, 14], "w": 1, "xyz": 1, "19": [1, 3], "15": [1, 7], "14": [1, 5], "7": [1, 3], "3": [1, 2, 3, 4, 8, 13], "output_fold": [1, 3, 5], "subexon_t": [1, 3, 10], "gene2speciesnam": 1, "connected_subexon": [1, 3], "none": [1, 3, 4, 5], "cutoff": [1, 3, 4], "30": [1, 3, 4, 13], "min_col_numb": [1, 3], "prographmsa": [1, 3, 7, 13], "pad": [1, 3, 13], "xxxxxxxxxx": [1, 3], "species_list": [1, 3, 4], "keep_single_subexon": [1, 3], "fals": [1, 3, 4, 6, 11, 13, 14], "return": [1, 2, 3, 4, 5, 6], "dict": [1, 3, 5], "data": [1, 2, 3, 4, 5, 6, 7, 10, 11], "tupl": [1, 3], "datafram": [1, 3, 4, 10], "chimer": [1, 3, 10, 13], "msa": [1, 3, 10], "take": [1, 3, 4, 10], "argument": [1, 2, 3, 4, 5, 7, 10, 12], "list": [1, 2, 3, 4, 5, 6, 10, 13, 14], "identifi": [1, 3, 10, 11, 13], "us": [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 13, 14], "given": [1, 4, 5], "all": [1, 3, 4, 5, 8, 10, 13, 14], "s_exon": [1, 3, 10], "minimum_len": [1, 3, 4], "coverage_cutoff": [1, 3, 4], "80": [1, 3, 4, 13], "percent_identity_cutoff": [1, 3, 4], "gap_open_penalti": [1, 3, 4], "gap_extend_penalti": [1, 3, 4], "movement": [1, 3], "true": [1, 3, 4, 5, 6], "perform": [1, 4, 5, 8], "almost": 1, "transcript_t": [1, 3, 10], "inform": [1, 3, 4, 5, 7, 10, 14], "accord": [1, 5, 10], "blosum50": 1, "matrix": [1, 3], "input_fold": 1, "max_tsl_level": [1, 4], "merg": [1, 3, 4], "pars": [1, 2, 4, 5], "command": [1, 2, 5, 7, 10], "line": [1, 2, 5, 7, 10, 13, 14], "argpars": [1, 2, 5], "parser": [1, 2, 5], "cluster2data": [1, 3], "updat": [1, 8], "ad": [1, 4, 11], "disentangl": 1, "homologi": 1, "relationship": [1, 5, 14], "user": [2, 11], "csv": [2, 3, 4, 5, 10, 11], "file": [2, 3, 4, 5, 10, 11, 13, 14], "previous": [2, 7, 11], "download": [2, 3, 4, 5, 7, 8, 10, 11, 14], "ensembl": [2, 3, 4, 5, 7, 10, 11, 13, 14], "class": [2, 3, 4], "properti": [2, 3], "alia": [2, 3], "field": [2, 3], "number": [2, 3, 4, 5, 10, 11, 13, 14], "input_df": 2, "seqrecord": [2, 3, 4], "modifi": [2, 3, 4, 11], "append": 2, "new": [2, 3, 4], "tsl_df": 2, "rais": [2, 3, 6], "error": [2, 3, 4, 6], "problem": 2, "arg": 2, "namedtupl": 2, "output": [2, 3, 4, 6, 7, 13], "sequences_path": 2, "fasta": [2, 3, 4, 5, 10], "creat": [3, 5, 6, 10, 11], "consensu": 3, "block": [3, 13], "margin": 3, "column": [3, 4, 5, 10, 11, 13], "msa_matrix": 3, "subexon_matrix": 3, "geneid": [3, 5, 11], "subexonindex": 3, "convert": 3, "cell": [3, 10], "index": [3, 4, 7], "nan": [3, 6], "gap": [3, 13], "binari": [3, 8], "show": [3, 10, 11, 13, 14], "presenc": 3, "seq": [3, 4, 6], "replac": [3, 4], "mhgl": 3, "xxxx": 3, "xxx": [3, 4], "xxxklmhsxxxxxxx": 3, "x": [3, 4, 8, 13], "klmh": 3, "mhglxklmhsx": 3, "subexon_df": 3, "subexonidclust": 3, "delet": [3, 4, 10, 11, 13], "msa_numpi": 3, "one": [3, 4, 8, 10, 11, 13, 14], "length": [3, 4, 6, 13], "contigu": 3, "transcript_data": 3, "threshold": 3, "appear": [3, 9], "more": [3, 7, 9, 13, 14], "than": [3, 4, 8, 10, 11, 13, 14], "fraction": [3, 10], "without": [3, 13, 14], "count": 3, "ani": [3, 13], "present": [3, 10], "introduc": 3, "indic": [3, 10, 11, 13, 14], "ambigu": 3, "50": [3, 4], "displai": [3, 7], "lowercas": [3, 11], "bio": 3, "alphabet": 3, "import": [3, 4, 6, 8], "iupac": 3, "multipleseqalign": 3, "seq_1": 3, "seeeeaccc": 3, "id": [3, 4, 5, 9, 10, 11, 14], "seq_2": 3, "sseeegcc": 3, "ii": [3, 7], "seq_3": 3, "ssseekc": 3, "iii": 3, "seq_4": 3, "sssseh": 3, "iiii": 3, "ssxeexccc": 3, "biopython": [3, 4], "col_clust": 3, "seq_id": 3, "full": 3, "key_column": 3, "valu": [3, 4, 10], "do": [3, 4, 8, 11, 13], "conform": [3, 6], "name": [3, 4, 5, 6, 7, 9, 10, 12], "0_number": 3, "row": [3, 4, 10, 13], "block_start": 3, "block_end": 3, "destination_start": 3, "destination_end": 3, "move": [3, 13], "place": [3, 4, 6], "numpi": [3, 6], "np": [3, 6], "mat": 3, "8": [3, 4], "9": [3, 8], "max_res_block": 3, "min_gap_block": 3, "find": [3, 4, 7, 9, 10, 11], "problemat": 3, "swap": 3, "within": [3, 4, 11], "its": [3, 4, 7, 8, 11], "depend": [3, 7], "e": [3, 4, 7, 8, 9, 10, 11, 13, 14], "g": [3, 4, 9, 10, 11, 13, 14], "last_block": 3, "first_block": 3, "gene_id": 3, "str": [3, 4, 5, 6], "check": [3, 4, 7, 9], "matric": 3, "separ": [3, 10, 13, 14], "rest": 3, "13": 3, "msa_fil": 3, "object": 3, "empti": [3, 4, 10], "matrix_row": 3, "three": [3, 9, 10], "output_path": 3, "run": [3, 7, 8, 10, 11, 13], "you": [3, 7, 8, 9, 10, 11], "pass": [3, 4], "default": [3, 4, 5, 10, 11, 13, 14], "mldist_gap": 3, "instal": [3, 7], "http": [3, 4, 8], "github": [3, 8, 11], "com": [3, 8], "acg": 3, "team": [3, 8], "have": [3, 4, 8, 9, 10, 11, 13], "ubuntu": [3, 8], "subsystem": [3, 8], "linux": [3, 8], "try": 3, "option": [3, 4, 7, 10], "ex": 3, "c": [3, 13], "bash": [3, 8], "wsl": 3, "save": [3, 5, 13], "about": [3, 4, 5, 10], "like": [3, 11], "ortholog": [3, 5, 9, 10, 13, 14], "region": [3, 9, 10, 11, 13, 14], "time": [3, 5, 11], "cross": 3, "boundari": [3, 4, 11], "gene2sp": 3, "sort": [3, 4, 13], "id_column": 3, "set": [3, 4, 5, 13], "connect": [3, 10], "element": [3, 4, 10], "direct": [3, 5], "pair": [3, 10], "befor": [3, 4, 11], "least": [3, 4, 5], "detect": [3, 10], "conserv": [3, 7, 10], "altern": [3, 7], "splice": [3, 4, 7, 9, 13], "event": [3, 7], "canon": [3, 10], "ASEs": 3, "s_exon_t": [3, 10], "s_exon_msa": 3, "graph_file_nam": 3, "min_gen": 3, "min_transcript": 3, "column_ord": 3, "delim": [3, 4], "A": [3, 4, 6, 7, 11], "ASE": [3, 10], "frame": [3, 4], "dictionari": [3, 5], "path_tabl": [3, 10, 13], "imput": 3, "panda": [3, 4, 10], "kei": [3, 5], "canonical_path": 3, "alternative_path": 3, "weight": [3, 10], "type": [3, 4, 5, 8], "insert": [3, 10], "mutual": [3, 10], "exclus": [3, 10], "non": [3, 4, 10, 11], "last": [3, 10, 11], "choos": 3, "read": [3, 4, 7, 10], "networkx": 3, "store": [3, 4, 5, 10], "five": 3, "node": [3, 10], "edg": [3, 10], "score": 3, "filenam": 3, "s_exons_msa": 3, "node2gen": 3, "edge2gen": 3, "node2transcript": 3, "edge2transcript": 3, "edge2trx_con": 3, "s_exon_2_char": 3, "gml": [3, 10], "model": [3, 4, 10], "languag": [3, 10], "format": [3, 4, 5, 6, 10], "featur": [3, 5], "gener": [3, 5, 10, 13], "singl": [3, 4, 10, 11, 13, 14], "result": [3, 7, 8, 9, 10], "1_0": 3, "1_1": 3, "2_0": 3, "s_exon2char": 3, "represent": [3, 4], "s_exon_data": 3, "ensembl_fold": 3, "input_tre": 3, "output_tre": 3, "exontable_fil": 3, "used_gen": 3, "unus": 3, "tree": [3, 5, 8, 10], "newick": 3, "prune": 3, "occur": [3, 5], "doe": 3, "except": [3, 10], "translat": 3, "longest": [3, 4, 10], "get": [3, 4, 5, 6, 8, 14], "map": [3, 4, 10], "j": [3, 10], "plotli": [3, 10, 13], "python": [3, 8, 10], "structur": [3, 4, 5, 6, 10], "javascript": 3, "html": [3, 10, 13], "seq_i": 3, "seq_j": 3, "ungap": 3, "should": [3, 4, 5, 10, 11, 13], "mhkclvde": 3, "ytedqggfrk": 3, "mllhyhhhkc": 3, "lmcytrdlhg": 3, "ih": 3, "l": [3, 5, 10, 13, 14], "k": 3, "mllhyh": 3, "ihlk": 3, "xxxxxxxxxxmhkclvde": 3, "xxlmcytrdlhg": 3, "xxlmc": 3, "subexon2clust": 3, "substitution_matrix": [3, 4], "execut": 3, "look": [3, 10, 14], "mean": [3, 10, 11], "been": [3, 11, 13, 14], "clean": 3, "up": 3, "against": [3, 4, 13], "other": [3, 4, 7, 10, 13], "differ": [3, 4, 6, 10, 11], "origin": [3, 13], "nice": 3, "anoth": 3, "match": [3, 4], "propos": 3, "rel": 3, "pairwis": [3, 4], "merge_non_redund": 3, "togeth": [3, 10], "onli": [3, 4, 9, 10], "joint": 3, "happen": 3, "reduc": 3, "redund": 3, "some": 3, "give": [3, 4, 10], "disjoint": 3, "produc": 3, "chr": 3, "report": 3, "well": 3, "call": [3, 5, 11], "inclus": 3, "20": [3, 4], "40": 3, "31": 3, "39": 3, "traceback": 3, "most": 3, "recent": 3, "assertionerror": 3, "to_merg": 3, "subexon_1": 3, "subexon_2": 3, "denorm": [3, 10], "final": [3, 8, 9, 10], "outputdir": [3, 13], "cut": 3, "helper": 4, "work": 4, "item2str": 4, "string": [4, 6, 10, 11], "100": 4, "b": 4, "lambda": 4, "element_i": 4, "element_j": 4, "fill": 4, "todelet": 4, "messag": 4, "warn": 4, "were": 4, "found": 4, "root": 4, "set_list": 4, "keep": 4, "seq_a": 4, "seq_b": 4, "sequnc": 4, "pairwise2": 4, "seq_len": 4, "aaaa": 4, "seq_queri": 4, "seq_target": 4, "shortest": 4, "aaaaaaaa": 4, "trx_data": 4, "after": [4, 8, 11], "local": [4, 8], "hobohm": 4, "algorithm": 4, "fast": 4, "order": [4, 9, 10, 11], "extra": 4, "queryexon": 4, "queri": 4, "join": 4, "smith": 4, "waterman": 4, "keyword": 4, "less": [4, 8, 11], "exonqueri": 4, "percent": [4, 13], "decid": 4, "until": 4, "thei": [4, 6, 11, 13], "reach": 4, "higher": [4, 8], "plu": 4, "allow": [4, 7, 8, 9, 10, 14], "low": 4, "better": 4, "ones": 4, "target": [4, 5, 8], "aa": 4, "chack": 4, "code": [4, 5, 8, 11], "wa": 4, "chang": [4, 8], "checker": 4, "could": [4, 13, 14], "help": [4, 8], "understand": 4, "xx": 4, "cdna_len": 4, "previous_end_phas": 4, "explain": 4, "glossari": 4, "www": 4, "org": 4, "multi": 4, "land": 4, "insid": 4, "1st": 4, "3rd": 4, "just": [4, 11], "which": 4, "under": [4, 14], "definit": 4, "actual": [4, 8], "y": [4, 13], "denot": 4, "No": 4, "interrupt": 4, "yyyzzz": 4, "xxyyyzzz": 4, "xyyyzzz": 4, "note": [4, 8, 11, 14], "xxxyyi": 4, "xyyyzz": 4, "xxyyyz": 4, "zz": 4, "data_fram": 4, "includ": [4, 10, 13, 14], "appli": 4, "must": 4, "stabl": [4, 5, 9, 14], "exonrank": [4, 11], "ensur": [4, 14], "throw": 4, "someth": [4, 11], "wrong": 4, "exon_po": 4, "prev_row_index": 4, "end_phase_column": 4, "endphas": [4, 11], "sequence_fil": 4, "exonsequ": 4, "descript": [4, 7, 11], "exonid": [4, 11], "ipt_in": 4, "allow_incomplete_cd": 4, "seq_column": 4, "exonproteinsequ": 4, "trx_column": 4, "transcriptid": [4, 11], "incomplet": [4, 11], "finish": 4, "probabl": 4, "cd": [4, 8, 10, 11], "different": 4, "fron": 4, "exon_id_column": 4, "similar": 4, "unifi": 4, "exon_table_fil": 4, "tsl_table_fil": 4, "exon_sequence_fil": 4, "remove_na": 4, "remove_badqu": 4, "integr": 4, "due": 4, "combin": 4, "primari": 4, "evid": 4, "filter": [4, 5, 13], "out": [4, 5, 7], "interest": 4, "our": [4, 7, 9], "analysi": 4, "ncrna": 4, "partial": 4, "etc": 4, "At": 4, "moment": 4, "biotyp": 4, "miss": 4, "elimin": 4, "summar": 4, "t": [4, 6, 10, 13], "whe": 4, "provid": [4, 5], "biomart": 4, "done": 4, "rank": [4, 10, 11], "process": 4, "peptid": 4, "account": [4, 10], "overlap": [4, 5], "alwai": [4, 11], "older": 4, "implement": 4, "tsl_file": 4, "support": [4, 6, 13], "determin": 4, "maximum": [4, 13], "junction": [4, 10], "suspect": 4, "mrna": 4, "best": 4, "flag": [4, 8], "multipl": [4, 5, 8, 9, 10, 11], "est": 4, "express": 4, "tag": 4, "multiindex": 4, "cluster_list": 4, "default_valu": 4, "column_nam": 4, "get_item": 4, "paramet": [4, 5], "exist": [4, 6], "containt": 4, "item": 4, "getitem": 4, "delimit": [4, 10], "pd": 4, "df": 4, "manag": 4, "fri": 5, "apr": 5, "16": 5, "32": [5, 8], "51": 5, "2017": 5, "author": 5, "huguesrichard": 5, "diegozea": [5, 8], "un": 5, "premier": 5, "jeu": 5, "de": 5, "fonction": 5, "pour": 5, "fair": 5, "requ\u00eat": 5, "avec": 5, "api": [5, 7], "restful": 5, "afin": 5, "r\u00e9cup\u00e9rer": 5, "tou": 5, "le": 5, "g\u00e8ne": 5, "homologu": 5, "\u00e0": 5, "donn\u00e9": 5, "par": 5, "son": 5, "nom": 5, "courant": 5, "dseq": 5, "write": 5, "dortho": 5, "n": [5, 14], "orthologu": 5, "extens": [5, 13], "param": 5, "header": 5, "request": 5, "species_nam": [5, 6], "listensexon": 5, "symbol": 5, "xref": 5, "human": [5, 9, 10], "mapk8": [5, 9, 10, 11, 14], "ensgeneid": 5, "around": 5, "now": 5, "whole": 5, "restrict": 5, "lortholog": 5, "wrapper": 5, "dn_d": 5, "float": 5, "method_link_typ": 5, "rop": 5, "sourc": [5, 7], "taxonomy_level": 5, "enum": 5, "ortholog_one2on": 5, "ortholog_one2mani": 5, "within_species_paralog": 5, "align_seq": 5, "perc_po": 5, "protein_id": 5, "perc_id": 5, "cigar_lin": 5, "taxon_id": 5, "int": 5, "form": [5, 10], "en": [5, 9], "prefix": 5, "uniqu": 5, "eleven": 5, "digit": 5, "listofdict": 5, "fname": 5, "stream": 5, "select": 5, "otherwis": [5, 11], "onc": 5, "written": 5, "ensembl_vers": [5, 10], "txt": [5, 10], "l_of_sptr": 5, "obtain": 5, "databas": [5, 10], "subfold": [6, 10, 13], "doesn": 6, "correct": 6, "expect": 6, "binomi": [6, 11], "trinomi": 6, "aren": 6, "don": 6, "homo_sapien": [6, 9, 10, 13, 14], "colobus_angolensis_palliatu": 6, "cricetulus_griseus_chok1gshd": 6, "specieslist": [6, 9, 13, 14], "mus_musculu": [6, 9, 10, 13, 14], "3_0": 6, "3_1": 6, "4_0": 6, "len": 6, "mlghac": 6, "mlgh": 6, "ac": 6, "test": [6, 11], "collect": 6, "assess": 7, "evolutionari": [7, 9], "youtub": [7, 9], "tutori": [7, 9], "_": 7, "see": [7, 9, 11], "how": [7, 9], "easili": [7, 10], "visualis": [7, 9], "want": [7, 8, 9, 11], "know": 7, "research": 7, "public": [7, 9], "your": [7, 9, 11], "pleas": [7, 8], "cite": 7, "u": [7, 10], "zea": 7, "dj": 7, "laskina": 7, "baudin": 7, "richard": 7, "h": [7, 9, 11, 13, 14], "lain": 7, "2021": 7, "jun": 7, "gr": 7, "274696": 7, "doi": 7, "1101": 7, "120": 7, "Ases": 7, "web": 7, "server": 7, "conveni": 7, "make": [7, 8], "easi": 7, "explor": [7, 8], "docker": 7, "imag": 7, "quick": 7, "introduct": 7, "program": [7, 8, 9, 10, 11, 13, 14], "prepar": [7, 12], "intermedi": 7, "search": 7, "page": 7, "instruct": 8, "link": 8, "compil": 8, "bit": 8, "64": 8, "osx": 8, "renam": 8, "put": 8, "through": 8, "wget": 8, "raw": 8, "master": 8, "bin": 8, "prographmsa_64": 8, "chmod": 8, "sudo": 8, "mv": 8, "easiest": 8, "pypi": 8, "pip": 8, "m": [8, 13, 14], "mai": 8, "python3": 8, "latest": 8, "develop": [8, 11, 13, 14], "clone": 8, "repo": 8, "git": 8, "pull": 8, "reflect": 8, "offer": [8, 9], "To": 8, "websit": 8, "powershel": 8, "directori": [8, 11, 13], "bind": 8, "mount": 8, "project": 8, "pwd": 8, "unix": 8, "ti": 8, "rm": 8, "access": 8, "termin": 8, "04": 8, "ipython3": 8, "easier": 8, "potenti": 9, "creation": 9, "esg": 9, "There": 9, "detail": [9, 11, 14], "toward": 9, "goal": 9, "meant": 9, "being": 9, "requir": [9, 11], "document": 9, "analys": 9, "mous": 9, "As": [9, 10], "synonym": 9, "clash": 9, "safer": 9, "ensg00000107643": [9, 11, 14], "avail": [9, 10, 13, 14], "experi": 9, "manual": 9, "them": 9, "configur": 9, "pax6": 10, "xenopus_tropicali": 10, "danio_rerio": 10, "That": 10, "four": 10, "exonst": 10, "tsv": 10, "nh": 10, "date": 10, "hour": 10, "assembli": 10, "ense00003837028": 10, "chromosom": 10, "grch38": 10, "48306753": 10, "48306821": 10, "cggcgaccaccccggacggcccctgtccccgctggcgggcttccctgtcgccgttcgctgcgctgccgg": 10, "entri": 10, "plot_chimer": [10, 13], "studi": 10, "_intermedi": [10, 13], "chimeric_alignment_1": 10, "cluster_data": 10, "cluster_plot": 10, "gene_ids_1": 10, "msa_matrix_1": 10, "subexon_table_1": 10, "ases_t": 10, "msa_s_exon_10_0": 10, "msa_s_exon_9_0": 10, "pir": 10, "splice_graph": 10, "seed": 10, "homolog": 10, "hmmsearch": 10, "rather": [10, 11], "normal": 10, "These": 10, "comma": [10, 13, 14], "r": [10, 13], "julia": 10, "among": 10, "softwar": 10, "tool": [10, 13], "re": 10, "strsplit": 10, "method": 10, "observ": 10, "concaten": 10, "isoform": 10, "automat": 10, "largest": 10, "highest": 10, "minim": 10, "transcript_weighted_conserv": 10, "compar": 10, "canonicalpath": 10, "alternativepath": 10, "subpath": 10, "classifi": 10, "fully_altern": 10, "alternative_start": 10, "alternative_end": 10, "simpli": 10, "classif": 10, "further": 10, "mutually_exclus": 10, "partially_mutually_exclus": 10, "mutualexclus": 10, "mutualexclusivecanon": 10, "mutualexclusivealtern": 10, "15_0": 10, "15_1": 10, "7_5": 10, "here": 10, "both": [10, 13], "consequ": 10, "necessarili": 10, "readabl": 10, "rich": 10, "metadata": 10, "label": 10, "transcript_fract": 10, "over": 10, "total": 10, "orthologi": [10, 14], "dataset": 10, "molecular": 10, "unicod": 10, "exoncluster_chimericblock": 10, "annot": 10, "state": 10, "interact": 10, "constitut": 10, "load": 10, "freez": 10, "tab": 10, "lot": 10, "comput": [10, 11, 13, 14], "hasn": 10, "enough": 10, "resourc": 10, "thorax": [11, 14], "describ": 11, "user_transcript": 11, "usag": [11, 13, 14], "verbos": [11, 14], "print": [11, 14], "progress": [11, 14], "exit": [11, 13, 14], "lcqb": [11, 13, 14], "laboratori": [11, 13, 14], "quantit": [11, 13, 14], "biologi": [11, 13, 14], "umr": [11, 13, 14], "7238": [11, 13, 14], "cnr": [11, 13, 14], "sorbonn": [11, 13, 14], "universit\u00e9": [11, 13, 14], "while": 11, "unless": 11, "would": 11, "underscor": 11, "space": [11, 13, 14], "forward": 11, "consecut": 11, "integ": 11, "exonregionstart": 11, "nucleotidesequ": 11, "exonregionend": 11, "genomiccodingstart": 11, "genomiccodingend": 11, "startphas": 11, "zero": 11, "fall": 11, "third": 11, "wherea": 11, "utr": 11, "complement": 11, "repositori": 11, "googl": 11, "spreadsheet": 11, "copi": 11, "inputdir": 13, "o": [13, 14], "maxtsl": 13, "minlen": 13, "mingen": 13, "mintranscript": 13, "p": 13, "gapopen": 13, "gapextend": 13, "no_mov": 13, "no_disintegr": 13, "canonical_criteria": 13, "minimum": 13, "consid": 13, "shorter": 13, "penalti": 13, "open": 13, "rescue_unaligned_subexon": 13, "reassign": 13, "dure": 13, "By": 13, "kept": 13, "per": [13, 14], "noth": [13, 14], "minimumconserv": 13, "minimumtranscriptweightedconserv": 13, "meantranscriptweightedconserv": 13, "transcriptlength": 13, "avoid": 13, "genenam": 14, "setup": 14}, "objects": {"": [[1, 0, 0, "-", "thoraxe"]], "thoraxe": [[2, 0, 0, "-", "add_transcripts"], [3, 0, 0, "-", "subexons"], [1, 0, 0, "-", "thoraxe"], [4, 0, 0, "-", "transcript_info"], [5, 0, 0, "-", "transcript_query"], [6, 0, 0, "-", "utils"], [1, 0, 0, "-", "version"]], "thoraxe.add_transcripts": [[2, 0, 0, "-", "add_transcripts"]], "thoraxe.add_transcripts.add_transcripts": [[2, 1, 1, "", "Paths"], [2, 3, 1, "", "add_sequences"], [2, 3, 1, "", "add_to_exontable"], [2, 3, 1, "", "add_to_tsl"], [2, 3, 1, "", "check_input"], [2, 3, 1, "", "get_output_paths"], [2, 3, 1, "", "main"], [2, 3, 1, "", "parse_command_line"], [2, 3, 1, "", "read_fasta"]], "thoraxe.add_transcripts.add_transcripts.Paths": [[2, 2, 1, "", "exontable"], [2, 2, 1, "", "input"], [2, 2, 1, "", "sequences"], [2, 2, 1, "", "tsl"]], "thoraxe.subexons": [[3, 0, 0, "-", "alignment"], [3, 0, 0, "-", "ases"], [3, 0, 0, "-", "graph"], [3, 0, 0, "-", "phylosofs"], [3, 0, 0, "-", "plot"], [3, 0, 0, "-", "rescue"], [3, 0, 0, "-", "subexons"], [3, 0, 0, "-", "tidy"]], "thoraxe.subexons.alignment": [[3, 1, 1, "", "ColCluster"], [3, 1, 1, "", "ColPattern"], [3, 1, 1, "", "ProblematicSubexonBlock"], [3, 3, 1, "", "cluster_subexon_blocks"], [3, 3, 1, "", "column_clusters"], [3, 3, 1, "", "column_patterns"], [3, 3, 1, "", "create_chimeric_sequences"], [3, 3, 1, "", "create_msa_matrix"], [3, 3, 1, "", "create_subexon_matrix"], [3, 3, 1, "", "delete_padding"], [3, 3, 1, "", "delete_subexons"], [3, 3, 1, "", "disintegration"], [3, 3, 1, "", "gene2species"], [3, 3, 1, "", "get_consensus"], [3, 3, 1, "", "get_gene_ids"], [3, 3, 1, "", "get_subexon_boundaries"], [3, 3, 1, "", "get_submsa"], [3, 3, 1, "", "impute_missing_s_exon"], [3, 3, 1, "", "move_block"], [3, 3, 1, "", "move_problematic_block_clusters"], [3, 3, 1, "", "move_subexon_block"], [3, 3, 1, "", "msa2sequences"], [3, 3, 1, "", "msa_matrices"], [3, 3, 1, "", "problematic_block_clusters"], [3, 3, 1, "", "problematic_subexon_blocks"], [3, 3, 1, "", "read_msa_fasta"], [3, 3, 1, "", "resume_seq"], [3, 3, 1, "", "run_aligner"], [3, 3, 1, "", "save_s_exons"], [3, 3, 1, "", "score_solution"], [3, 3, 1, "", "sort_species"], [3, 3, 1, "", "subexon_connectivity"]], "thoraxe.subexons.alignment.ColPattern": [[3, 2, 1, "", "end"], [3, 2, 1, "", "pattern"], [3, 2, 1, "", "start"]], "thoraxe.subexons.alignment.ProblematicSubexonBlock": [[3, 2, 1, "", "block_type"], [3, 2, 1, "", "gap_block_end"], [3, 2, 1, "", "gap_block_start"], [3, 2, 1, "", "sequence_index"], [3, 2, 1, "", "subexon"], [3, 2, 1, "", "subexon_block_end"], [3, 2, 1, "", "subexon_block_start"], [3, 2, 1, "", "subexon_blocks"]], "thoraxe.subexons.ases": [[3, 1, 1, "", "PathData"], [3, 3, 1, "", "conserved_ases"], [3, 3, 1, "", "create_ases_table"], [3, 3, 1, "", "detect_ases"], [3, 3, 1, "", "get_transcript_scores"], [3, 3, 1, "", "read_splice_graph"]], "thoraxe.subexons.ases.PathData": [[3, 2, 1, "", "Genes"], [3, 2, 1, "", "Transcripts"]], "thoraxe.subexons.graph": [[3, 3, 1, "", "nodes_and_edges2genes_and_transcripts"], [3, 3, 1, "", "splice_graph_gml"]], "thoraxe.subexons.phylosofs": [[3, 3, 1, "", "get_s_exon2char"], [3, 3, 1, "", "get_transcript2phylosofs"], [3, 3, 1, "", "phylosofs_inputs"], [3, 3, 1, "", "prune_tree"]], "thoraxe.subexons.plot": [[3, 3, 1, "", "create_python_structure"], [3, 3, 1, "", "plot_msa_subexons"]], "thoraxe.subexons.rescue": [[3, 3, 1, "", "get_unaligned_regions"], [3, 3, 1, "", "modify_subexon_cluster"], [3, 3, 1, "", "subexon_rescue_phase"]], "thoraxe.subexons.subexons": [[3, 1, 1, "", "Interval"], [3, 3, 1, "", "create_subexon_table"], [3, 3, 1, "", "disjoint_intervals"], [3, 3, 1, "", "update_to_merge_list"]], "thoraxe.subexons.subexons.Interval": [[3, 2, 1, "", "components"], [3, 2, 1, "", "end"], [3, 2, 1, "", "start"]], "thoraxe.subexons.tidy": [[3, 3, 1, "", "get_tidy_table"]], "thoraxe.thoraxe": [[1, 3, 1, "", "add_s_exon_phases_and_coordinates"], [1, 3, 1, "", "create_chimeric_msa"], [1, 3, 1, "", "get_s_exon_msas"], [1, 3, 1, "", "get_s_exons"], [1, 3, 1, "", "get_subexons"], [1, 3, 1, "", "get_transcripts"], [1, 3, 1, "", "main"], [1, 3, 1, "", "merge_clusters"], [1, 3, 1, "", "parse_command_line"], [1, 3, 1, "", "update_subexon_table"]], "thoraxe.transcript_info": [[4, 0, 0, "-", "clusters"], [4, 0, 0, "-", "exon_clustering"], [4, 0, 0, "-", "phases"], [4, 0, 0, "-", "transcript_info"]], "thoraxe.transcript_info.clusters": [[4, 3, 1, "", "cluster2str"], [4, 3, 1, "", "fill_clusters"], [4, 3, 1, "", "inform_about_deletions"], [4, 3, 1, "", "set_to_delete"]], "thoraxe.transcript_info.exon_clustering": [[4, 3, 1, "", "align"], [4, 3, 1, "", "coverage"], [4, 3, 1, "", "coverage_shortest"], [4, 3, 1, "", "exon_clustering"], [4, 3, 1, "", "percent_identity"]], "thoraxe.transcript_info.phases": [[4, 3, 1, "", "bases_to_complete_next_codon"], [4, 3, 1, "", "bases_to_complete_previous_codon"], [4, 3, 1, "", "calculate_phase"], [4, 3, 1, "", "check_order_and_phases"], [4, 3, 1, "", "end_phase_previous_exon"]], "thoraxe.transcript_info.transcript_info": [[4, 3, 1, "", "add_exon_sequences"], [4, 3, 1, "", "add_protein_seq"], [4, 3, 1, "", "delete_badquality_sequences"], [4, 3, 1, "", "delete_identical_sequences"], [4, 3, 1, "", "delete_incomplete_sequences"], [4, 3, 1, "", "find_identical_exons"], [4, 3, 1, "", "find_identical_sequences"], [4, 3, 1, "", "merge_identical_exons"], [4, 3, 1, "", "read_exon_file"], [4, 3, 1, "", "read_transcript_info"], [4, 3, 1, "", "read_tsl_file"], [4, 3, 1, "", "store_cluster"]], "thoraxe.transcript_query": [[5, 0, 0, "-", "transcript_query"]], "thoraxe.transcript_query.transcript_query": [[5, 3, 1, "", "dictseq2fasta"], [5, 3, 1, "", "filter_ortho"], [5, 3, 1, "", "generic_ensembl_rest_request"], [5, 3, 1, "", "get_biomart_exons_annot"], [5, 3, 1, "", "get_exons_sequences"], [5, 3, 1, "", "get_geneids_from_symbol"], [5, 3, 1, "", "get_genetree"], [5, 3, 1, "", "get_listofexons"], [5, 3, 1, "", "get_listoftranscripts"], [5, 3, 1, "", "get_orthologs"], [5, 3, 1, "", "get_transcripts_orthologs"], [5, 3, 1, "", "is_esemble_id"], [5, 3, 1, "", "lodict2csv"], [5, 3, 1, "", "main"], [5, 3, 1, "", "parse_command_line"], [5, 3, 1, "", "save_ensembl_version"], [5, 3, 1, "", "write_tsl_file"]], "thoraxe.utils": [[6, 0, 0, "-", "folders"], [6, 0, 0, "-", "species"], [6, 0, 0, "-", "split"], [6, 0, 0, "-", "windows"]], "thoraxe.utils.folders": [[6, 3, 1, "", "create_subfolder"]], "thoraxe.utils.species": [[6, 3, 1, "", "check_species_name"], [6, 3, 1, "", "get_species_list"]], "thoraxe.utils.split": [[6, 3, 1, "", "split_exons"], [6, 3, 1, "", "split_lengths"], [6, 3, 1, "", "split_seqs"]], "thoraxe.utils.windows": [[6, 3, 1, "", "is_windows"]]}, "objtypes": {"0": "py:module", "1": "py:class", "2": "py:property", "3": "py:function"}, "objnames": {"0": ["py", "module", "Python module"], "1": ["py", "class", "Python class"], "2": ["py", "property", "Python property"], "3": ["py", "function", "Python function"]}, "titleterms": {"thorax": [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 12, 13], "packag": [1, 2, 3, 4, 5, 6], "subpackag": 1, "submodul": [1, 2, 3, 4, 5, 6], "modul": [1, 2, 3, 4, 5, 6], "version": 1, "content": [1, 2, 3, 4, 5, 6], "add_transcript": [2, 11], "subexon": 3, "align": 3, "ases": 3, "graph": [3, 10], "phylosof": [3, 10], "plot": 3, "rescu": 3, "tidi": 3, "transcript_info": 4, "cluster": 4, "exon_clust": 4, "phase": 4, "transcript_queri": [5, 10, 14], "util": 6, "folder": 6, "speci": 6, "split": 6, "window": 6, "": [7, 10], "document": 7, "indic": 7, "tabl": [7, 10], "instal": 8, "depend": 8, "prographmsa": 8, "from": [8, 9], "sourc": 8, "docker": 8, "imag": 8, "quick": 9, "introduct": 9, "i": 9, "download": 9, "transcript": 9, "inform": 9, "ensembl": 9, "option": 9, "add": 9, "previous": 9, "data": 9, "ii": 9, "run": 9, "output": 10, "descript": 10, "exon": 10, "path": 10, "altern": 10, "splice": 10, "event": 10, "intermedi": 10, "posit": [11, 14], "argument": [11, 13, 14], "name": [11, 13, 14], "input": 11, "prepar": 11}, "envversion": {"sphinx.domains.c": 2, "sphinx.domains.changeset": 1, "sphinx.domains.citation": 1, "sphinx.domains.cpp": 8, "sphinx.domains.index": 1, "sphinx.domains.javascript": 2, "sphinx.domains.math": 2, "sphinx.domains.python": 3, "sphinx.domains.rst": 2, "sphinx.domains.std": 2, "sphinx": 57}, "alltitles": {"thoraxe": [[0, "thoraxe"], [10, "thoraxe"], [12, "thoraxe"], [13, "thoraxe"]], "thoraxe package": [[1, "thoraxe-package"]], "Subpackages": [[1, "subpackages"]], "Submodules": [[1, "submodules"], [2, "submodules"], [3, "submodules"], [4, "submodules"], [5, "submodules"], [6, "submodules"]], "thoraxe.thoraxe module": [[1, "module-thoraxe.thoraxe"]], "thoraxe.version module": [[1, "module-thoraxe.version"]], "Module contents": [[1, "module-thoraxe"], [2, "module-thoraxe.add_transcripts"], [3, "module-thoraxe.subexons"], [4, "module-thoraxe.transcript_info"], [5, "module-thoraxe.transcript_query"], [6, "module-thoraxe.utils"]], "thoraxe.add_transcripts package": [[2, "thoraxe-add-transcripts-package"]], "thoraxe.add_transcripts.add_transcripts module": [[2, "module-thoraxe.add_transcripts.add_transcripts"]], "thoraxe.subexons package": [[3, "thoraxe-subexons-package"]], "thoraxe.subexons.alignment module": [[3, "module-thoraxe.subexons.alignment"]], "thoraxe.subexons.ases module": [[3, "module-thoraxe.subexons.ases"]], "thoraxe.subexons.graph module": [[3, "module-thoraxe.subexons.graph"]], "thoraxe.subexons.phylosofs module": [[3, "module-thoraxe.subexons.phylosofs"]], "thoraxe.subexons.plot module": [[3, "module-thoraxe.subexons.plot"]], "thoraxe.subexons.rescue module": [[3, "module-thoraxe.subexons.rescue"]], "thoraxe.subexons.subexons module": [[3, "module-thoraxe.subexons.subexons"]], "thoraxe.subexons.tidy module": [[3, "module-thoraxe.subexons.tidy"]], "thoraxe.transcript_info package": [[4, "thoraxe-transcript-info-package"]], "thoraxe.transcript_info.clusters module": [[4, "module-thoraxe.transcript_info.clusters"]], "thoraxe.transcript_info.exon_clustering module": [[4, "module-thoraxe.transcript_info.exon_clustering"]], "thoraxe.transcript_info.phases module": [[4, "module-thoraxe.transcript_info.phases"]], "thoraxe.transcript_info.transcript_info module": [[4, "module-thoraxe.transcript_info.transcript_info"]], "thoraxe.transcript_query package": [[5, "thoraxe-transcript-query-package"]], "thoraxe.transcript_query.transcript_query module": [[5, "module-thoraxe.transcript_query.transcript_query"]], "thoraxe.utils package": [[6, "thoraxe-utils-package"]], "thoraxe.utils.folders module": [[6, "module-thoraxe.utils.folders"]], "thoraxe.utils.species module": [[6, "module-thoraxe.utils.species"]], "thoraxe.utils.split module": [[6, "module-thoraxe.utils.split"]], "thoraxe.utils.windows module": [[6, "module-thoraxe.utils.windows"]], "ThorAxe\u2019s documentation!": [[7, "thoraxe-s-documentation"]], "Indices and tables": [[7, "indices-and-tables"]], "Installation": [[8, "installation"]], "Dependencies": [[8, "dependencies"]], "ProGraphMSA": [[8, "prographmsa"]], "ThorAxe": [[8, "thoraxe"]], "Installation from source": [[8, "installation-from-source"]], "Docker image": [[8, "docker-image"]], "Quick Introduction": [[9, "quick-introduction"]], "I. Download transcript information from Ensembl": [[9, "i-download-transcript-information-from-ensembl"]], "[OPTIONAL] Add transcripts to the previously download data": [[9, "optional-add-transcripts-to-the-previously-download-data"]], "II. Run ThorAxe": [[9, "ii-run-thoraxe"]], "Output description": [[10, "output-description"]], "transcript_query": [[10, "transcript-query"], [14, "transcript-query"]], "Tables": [[10, "tables"]], "S-exon table": [[10, "s-exon-table"]], "Path table": [[10, "path-table"]], "Table of Alternative Splicing Events": [[10, "table-of-alternative-splicing-events"]], "Splice graph": [[10, "splice-graph"]], "phylosofs": [[10, "phylosofs"]], "Intermediate outputs": [[10, "intermediate-outputs"]], "add_transcripts": [[11, "add-transcripts"]], "Positional Arguments": [[11, "Positional Arguments"], [14, "Positional Arguments"]], "Named Arguments": [[11, "Named Arguments"], [13, "Named Arguments"], [14, "Named Arguments"]], "Input preparation": [[11, "input-preparation"]]}, "indexentries": {"add_s_exon_phases_and_coordinates() (in module thoraxe.thoraxe)": [[1, "thoraxe.thoraxe.add_s_exon_phases_and_coordinates"]], "create_chimeric_msa() (in module thoraxe.thoraxe)": [[1, "thoraxe.thoraxe.create_chimeric_msa"]], "get_s_exon_msas() (in module thoraxe.thoraxe)": [[1, "thoraxe.thoraxe.get_s_exon_msas"]], "get_s_exons() (in module thoraxe.thoraxe)": [[1, "thoraxe.thoraxe.get_s_exons"]], "get_subexons() (in module thoraxe.thoraxe)": [[1, "thoraxe.thoraxe.get_subexons"]], "get_transcripts() (in module thoraxe.thoraxe)": [[1, "thoraxe.thoraxe.get_transcripts"]], "main() (in module thoraxe.thoraxe)": [[1, "thoraxe.thoraxe.main"]], "merge_clusters() (in module thoraxe.thoraxe)": [[1, "thoraxe.thoraxe.merge_clusters"]], "module": [[1, "module-thoraxe"], [1, "module-thoraxe.thoraxe"], [1, "module-thoraxe.version"], [2, "module-thoraxe.add_transcripts"], [2, "module-thoraxe.add_transcripts.add_transcripts"], [3, "module-thoraxe.subexons"], [3, "module-thoraxe.subexons.alignment"], [3, "module-thoraxe.subexons.ases"], [3, "module-thoraxe.subexons.graph"], [3, "module-thoraxe.subexons.phylosofs"], [3, "module-thoraxe.subexons.plot"], [3, "module-thoraxe.subexons.rescue"], [3, "module-thoraxe.subexons.subexons"], [3, "module-thoraxe.subexons.tidy"], [4, "module-thoraxe.transcript_info"], [4, "module-thoraxe.transcript_info.clusters"], [4, "module-thoraxe.transcript_info.exon_clustering"], [4, "module-thoraxe.transcript_info.phases"], [4, "module-thoraxe.transcript_info.transcript_info"], [5, "module-thoraxe.transcript_query"], [5, "module-thoraxe.transcript_query.transcript_query"], [6, "module-thoraxe.utils"], [6, "module-thoraxe.utils.folders"], [6, "module-thoraxe.utils.species"], [6, "module-thoraxe.utils.split"], [6, "module-thoraxe.utils.windows"]], "parse_command_line() (in module thoraxe.thoraxe)": [[1, "thoraxe.thoraxe.parse_command_line"]], "thoraxe": [[1, "module-thoraxe"]], "thoraxe.thoraxe": [[1, "module-thoraxe.thoraxe"]], "thoraxe.version": [[1, "module-thoraxe.version"]], "update_subexon_table() (in module thoraxe.thoraxe)": [[1, "thoraxe.thoraxe.update_subexon_table"]], "paths (class in thoraxe.add_transcripts.add_transcripts)": [[2, "thoraxe.add_transcripts.add_transcripts.Paths"]], "add_sequences() (in module thoraxe.add_transcripts.add_transcripts)": [[2, "thoraxe.add_transcripts.add_transcripts.add_sequences"]], "add_to_exontable() (in module thoraxe.add_transcripts.add_transcripts)": [[2, "thoraxe.add_transcripts.add_transcripts.add_to_exontable"]], "add_to_tsl() (in module thoraxe.add_transcripts.add_transcripts)": [[2, "thoraxe.add_transcripts.add_transcripts.add_to_tsl"]], "check_input() (in module thoraxe.add_transcripts.add_transcripts)": [[2, "thoraxe.add_transcripts.add_transcripts.check_input"]], "exontable (thoraxe.add_transcripts.add_transcripts.paths property)": [[2, "thoraxe.add_transcripts.add_transcripts.Paths.exontable"]], "get_output_paths() (in module thoraxe.add_transcripts.add_transcripts)": [[2, "thoraxe.add_transcripts.add_transcripts.get_output_paths"]], "input (thoraxe.add_transcripts.add_transcripts.paths property)": [[2, "thoraxe.add_transcripts.add_transcripts.Paths.input"]], "main() (in module thoraxe.add_transcripts.add_transcripts)": [[2, "thoraxe.add_transcripts.add_transcripts.main"]], "parse_command_line() (in module thoraxe.add_transcripts.add_transcripts)": [[2, "thoraxe.add_transcripts.add_transcripts.parse_command_line"]], "read_fasta() (in module thoraxe.add_transcripts.add_transcripts)": [[2, "thoraxe.add_transcripts.add_transcripts.read_fasta"]], "sequences (thoraxe.add_transcripts.add_transcripts.paths property)": [[2, "thoraxe.add_transcripts.add_transcripts.Paths.sequences"]], "thoraxe.add_transcripts": [[2, "module-thoraxe.add_transcripts"]], "thoraxe.add_transcripts.add_transcripts": [[2, "module-thoraxe.add_transcripts.add_transcripts"]], "tsl (thoraxe.add_transcripts.add_transcripts.paths property)": [[2, "thoraxe.add_transcripts.add_transcripts.Paths.tsl"]], "colcluster (class in thoraxe.subexons.alignment)": [[3, "thoraxe.subexons.alignment.ColCluster"]], "colpattern (class in thoraxe.subexons.alignment)": [[3, "thoraxe.subexons.alignment.ColPattern"]], "genes (thoraxe.subexons.ases.pathdata property)": [[3, "thoraxe.subexons.ases.PathData.Genes"]], "interval (class in thoraxe.subexons.subexons)": [[3, "thoraxe.subexons.subexons.Interval"]], "pathdata (class in thoraxe.subexons.ases)": [[3, "thoraxe.subexons.ases.PathData"]], "problematicsubexonblock (class in thoraxe.subexons.alignment)": [[3, "thoraxe.subexons.alignment.ProblematicSubexonBlock"]], "transcripts (thoraxe.subexons.ases.pathdata property)": [[3, "thoraxe.subexons.ases.PathData.Transcripts"]], "block_type (thoraxe.subexons.alignment.problematicsubexonblock property)": [[3, "thoraxe.subexons.alignment.ProblematicSubexonBlock.block_type"]], "cluster_subexon_blocks() (in module thoraxe.subexons.alignment)": [[3, "thoraxe.subexons.alignment.cluster_subexon_blocks"]], "column_clusters() (in module thoraxe.subexons.alignment)": [[3, "thoraxe.subexons.alignment.column_clusters"]], "column_patterns() (in module thoraxe.subexons.alignment)": [[3, "thoraxe.subexons.alignment.column_patterns"]], "components (thoraxe.subexons.subexons.interval property)": [[3, "thoraxe.subexons.subexons.Interval.components"]], "conserved_ases() (in module thoraxe.subexons.ases)": [[3, "thoraxe.subexons.ases.conserved_ases"]], "create_ases_table() (in module thoraxe.subexons.ases)": [[3, "thoraxe.subexons.ases.create_ases_table"]], "create_chimeric_sequences() (in module thoraxe.subexons.alignment)": [[3, "thoraxe.subexons.alignment.create_chimeric_sequences"]], "create_msa_matrix() (in module thoraxe.subexons.alignment)": [[3, "thoraxe.subexons.alignment.create_msa_matrix"]], "create_python_structure() (in module thoraxe.subexons.plot)": [[3, "thoraxe.subexons.plot.create_python_structure"]], "create_subexon_matrix() (in module thoraxe.subexons.alignment)": [[3, "thoraxe.subexons.alignment.create_subexon_matrix"]], "create_subexon_table() (in module thoraxe.subexons.subexons)": [[3, "thoraxe.subexons.subexons.create_subexon_table"]], "delete_padding() (in module thoraxe.subexons.alignment)": [[3, "thoraxe.subexons.alignment.delete_padding"]], "delete_subexons() (in module thoraxe.subexons.alignment)": [[3, "thoraxe.subexons.alignment.delete_subexons"]], "detect_ases() (in module thoraxe.subexons.ases)": [[3, "thoraxe.subexons.ases.detect_ases"]], "disintegration() (in module thoraxe.subexons.alignment)": [[3, "thoraxe.subexons.alignment.disintegration"]], "disjoint_intervals() (in module thoraxe.subexons.subexons)": [[3, "thoraxe.subexons.subexons.disjoint_intervals"]], "end (thoraxe.subexons.alignment.colpattern property)": [[3, "thoraxe.subexons.alignment.ColPattern.end"]], "end (thoraxe.subexons.subexons.interval property)": [[3, "thoraxe.subexons.subexons.Interval.end"]], "gap_block_end (thoraxe.subexons.alignment.problematicsubexonblock property)": [[3, "thoraxe.subexons.alignment.ProblematicSubexonBlock.gap_block_end"]], "gap_block_start (thoraxe.subexons.alignment.problematicsubexonblock property)": [[3, "thoraxe.subexons.alignment.ProblematicSubexonBlock.gap_block_start"]], "gene2species() (in module thoraxe.subexons.alignment)": [[3, "thoraxe.subexons.alignment.gene2species"]], "get_consensus() (in module thoraxe.subexons.alignment)": [[3, "thoraxe.subexons.alignment.get_consensus"]], "get_gene_ids() (in module thoraxe.subexons.alignment)": [[3, "thoraxe.subexons.alignment.get_gene_ids"]], "get_s_exon2char() (in module thoraxe.subexons.phylosofs)": [[3, "thoraxe.subexons.phylosofs.get_s_exon2char"]], "get_subexon_boundaries() (in module thoraxe.subexons.alignment)": [[3, "thoraxe.subexons.alignment.get_subexon_boundaries"]], "get_submsa() (in module thoraxe.subexons.alignment)": [[3, "thoraxe.subexons.alignment.get_submsa"]], "get_tidy_table() (in module thoraxe.subexons.tidy)": [[3, "thoraxe.subexons.tidy.get_tidy_table"]], "get_transcript2phylosofs() (in module thoraxe.subexons.phylosofs)": [[3, "thoraxe.subexons.phylosofs.get_transcript2phylosofs"]], "get_transcript_scores() (in module thoraxe.subexons.ases)": [[3, "thoraxe.subexons.ases.get_transcript_scores"]], "get_unaligned_regions() (in module thoraxe.subexons.rescue)": [[3, "thoraxe.subexons.rescue.get_unaligned_regions"]], "impute_missing_s_exon() (in module thoraxe.subexons.alignment)": [[3, "thoraxe.subexons.alignment.impute_missing_s_exon"]], "modify_subexon_cluster() (in module thoraxe.subexons.rescue)": [[3, "thoraxe.subexons.rescue.modify_subexon_cluster"]], "move_block() (in module thoraxe.subexons.alignment)": [[3, "thoraxe.subexons.alignment.move_block"]], "move_problematic_block_clusters() (in module thoraxe.subexons.alignment)": [[3, "thoraxe.subexons.alignment.move_problematic_block_clusters"]], "move_subexon_block() (in module thoraxe.subexons.alignment)": [[3, "thoraxe.subexons.alignment.move_subexon_block"]], "msa2sequences() (in module thoraxe.subexons.alignment)": [[3, "thoraxe.subexons.alignment.msa2sequences"]], "msa_matrices() (in module thoraxe.subexons.alignment)": [[3, "thoraxe.subexons.alignment.msa_matrices"]], "nodes_and_edges2genes_and_transcripts() (in module thoraxe.subexons.graph)": [[3, "thoraxe.subexons.graph.nodes_and_edges2genes_and_transcripts"]], "pattern (thoraxe.subexons.alignment.colpattern property)": [[3, "thoraxe.subexons.alignment.ColPattern.pattern"]], "phylosofs_inputs() (in module thoraxe.subexons.phylosofs)": [[3, "thoraxe.subexons.phylosofs.phylosofs_inputs"]], "plot_msa_subexons() (in module thoraxe.subexons.plot)": [[3, "thoraxe.subexons.plot.plot_msa_subexons"]], "problematic_block_clusters() (in module thoraxe.subexons.alignment)": [[3, "thoraxe.subexons.alignment.problematic_block_clusters"]], "problematic_subexon_blocks() (in module thoraxe.subexons.alignment)": [[3, "thoraxe.subexons.alignment.problematic_subexon_blocks"]], "prune_tree() (in module thoraxe.subexons.phylosofs)": [[3, "thoraxe.subexons.phylosofs.prune_tree"]], "read_msa_fasta() (in module thoraxe.subexons.alignment)": [[3, "thoraxe.subexons.alignment.read_msa_fasta"]], "read_splice_graph() (in module thoraxe.subexons.ases)": [[3, "thoraxe.subexons.ases.read_splice_graph"]], "resume_seq() (in module thoraxe.subexons.alignment)": [[3, "thoraxe.subexons.alignment.resume_seq"]], "run_aligner() (in module thoraxe.subexons.alignment)": [[3, "thoraxe.subexons.alignment.run_aligner"]], "save_s_exons() (in module thoraxe.subexons.alignment)": [[3, "thoraxe.subexons.alignment.save_s_exons"]], "score_solution() (in module thoraxe.subexons.alignment)": [[3, "thoraxe.subexons.alignment.score_solution"]], "sequence_index (thoraxe.subexons.alignment.problematicsubexonblock property)": [[3, "thoraxe.subexons.alignment.ProblematicSubexonBlock.sequence_index"]], "sort_species() (in module thoraxe.subexons.alignment)": [[3, "thoraxe.subexons.alignment.sort_species"]], "splice_graph_gml() (in module thoraxe.subexons.graph)": [[3, "thoraxe.subexons.graph.splice_graph_gml"]], "start (thoraxe.subexons.alignment.colpattern property)": [[3, "thoraxe.subexons.alignment.ColPattern.start"]], "start (thoraxe.subexons.subexons.interval property)": [[3, "thoraxe.subexons.subexons.Interval.start"]], "subexon (thoraxe.subexons.alignment.problematicsubexonblock property)": [[3, "thoraxe.subexons.alignment.ProblematicSubexonBlock.subexon"]], "subexon_block_end (thoraxe.subexons.alignment.problematicsubexonblock property)": [[3, "thoraxe.subexons.alignment.ProblematicSubexonBlock.subexon_block_end"]], "subexon_block_start (thoraxe.subexons.alignment.problematicsubexonblock property)": [[3, "thoraxe.subexons.alignment.ProblematicSubexonBlock.subexon_block_start"]], "subexon_blocks (thoraxe.subexons.alignment.problematicsubexonblock property)": [[3, "thoraxe.subexons.alignment.ProblematicSubexonBlock.subexon_blocks"]], "subexon_connectivity() (in module thoraxe.subexons.alignment)": [[3, "thoraxe.subexons.alignment.subexon_connectivity"]], "subexon_rescue_phase() (in module thoraxe.subexons.rescue)": [[3, "thoraxe.subexons.rescue.subexon_rescue_phase"]], "thoraxe.subexons": [[3, "module-thoraxe.subexons"]], "thoraxe.subexons.alignment": [[3, "module-thoraxe.subexons.alignment"]], "thoraxe.subexons.ases": [[3, "module-thoraxe.subexons.ases"]], "thoraxe.subexons.graph": [[3, "module-thoraxe.subexons.graph"]], "thoraxe.subexons.phylosofs": [[3, "module-thoraxe.subexons.phylosofs"]], "thoraxe.subexons.plot": [[3, "module-thoraxe.subexons.plot"]], "thoraxe.subexons.rescue": [[3, "module-thoraxe.subexons.rescue"]], "thoraxe.subexons.subexons": [[3, "module-thoraxe.subexons.subexons"]], "thoraxe.subexons.tidy": [[3, "module-thoraxe.subexons.tidy"]], "update_to_merge_list() (in module thoraxe.subexons.subexons)": [[3, "thoraxe.subexons.subexons.update_to_merge_list"]], "add_exon_sequences() (in module thoraxe.transcript_info.transcript_info)": [[4, "thoraxe.transcript_info.transcript_info.add_exon_sequences"]], "add_protein_seq() (in module thoraxe.transcript_info.transcript_info)": [[4, "thoraxe.transcript_info.transcript_info.add_protein_seq"]], "align() (in module thoraxe.transcript_info.exon_clustering)": [[4, "thoraxe.transcript_info.exon_clustering.align"]], "bases_to_complete_next_codon() (in module thoraxe.transcript_info.phases)": [[4, "thoraxe.transcript_info.phases.bases_to_complete_next_codon"]], "bases_to_complete_previous_codon() (in module thoraxe.transcript_info.phases)": [[4, "thoraxe.transcript_info.phases.bases_to_complete_previous_codon"]], "calculate_phase() (in module thoraxe.transcript_info.phases)": [[4, "thoraxe.transcript_info.phases.calculate_phase"]], "check_order_and_phases() (in module thoraxe.transcript_info.phases)": [[4, "thoraxe.transcript_info.phases.check_order_and_phases"]], "cluster2str() (in module thoraxe.transcript_info.clusters)": [[4, "thoraxe.transcript_info.clusters.cluster2str"]], "coverage() (in module thoraxe.transcript_info.exon_clustering)": [[4, "thoraxe.transcript_info.exon_clustering.coverage"]], "coverage_shortest() (in module thoraxe.transcript_info.exon_clustering)": [[4, "thoraxe.transcript_info.exon_clustering.coverage_shortest"]], "delete_badquality_sequences() (in module thoraxe.transcript_info.transcript_info)": [[4, "thoraxe.transcript_info.transcript_info.delete_badquality_sequences"]], "delete_identical_sequences() (in module thoraxe.transcript_info.transcript_info)": [[4, "thoraxe.transcript_info.transcript_info.delete_identical_sequences"]], "delete_incomplete_sequences() (in module thoraxe.transcript_info.transcript_info)": [[4, "thoraxe.transcript_info.transcript_info.delete_incomplete_sequences"]], "end_phase_previous_exon() (in module thoraxe.transcript_info.phases)": [[4, "thoraxe.transcript_info.phases.end_phase_previous_exon"]], "exon_clustering() (in module thoraxe.transcript_info.exon_clustering)": [[4, "thoraxe.transcript_info.exon_clustering.exon_clustering"]], "fill_clusters() (in module thoraxe.transcript_info.clusters)": [[4, "thoraxe.transcript_info.clusters.fill_clusters"]], "find_identical_exons() (in module thoraxe.transcript_info.transcript_info)": [[4, "thoraxe.transcript_info.transcript_info.find_identical_exons"]], "find_identical_sequences() (in module thoraxe.transcript_info.transcript_info)": [[4, "thoraxe.transcript_info.transcript_info.find_identical_sequences"]], "inform_about_deletions() (in module thoraxe.transcript_info.clusters)": [[4, "thoraxe.transcript_info.clusters.inform_about_deletions"]], "merge_identical_exons() (in module thoraxe.transcript_info.transcript_info)": [[4, "thoraxe.transcript_info.transcript_info.merge_identical_exons"]], "percent_identity() (in module thoraxe.transcript_info.exon_clustering)": [[4, "thoraxe.transcript_info.exon_clustering.percent_identity"]], "read_exon_file() (in module thoraxe.transcript_info.transcript_info)": [[4, "thoraxe.transcript_info.transcript_info.read_exon_file"]], "read_transcript_info() (in module thoraxe.transcript_info.transcript_info)": [[4, "thoraxe.transcript_info.transcript_info.read_transcript_info"]], "read_tsl_file() (in module thoraxe.transcript_info.transcript_info)": [[4, "thoraxe.transcript_info.transcript_info.read_tsl_file"]], "set_to_delete() (in module thoraxe.transcript_info.clusters)": [[4, "thoraxe.transcript_info.clusters.set_to_delete"]], "store_cluster() (in module thoraxe.transcript_info.transcript_info)": [[4, "thoraxe.transcript_info.transcript_info.store_cluster"]], "thoraxe.transcript_info": [[4, "module-thoraxe.transcript_info"]], "thoraxe.transcript_info.clusters": [[4, "module-thoraxe.transcript_info.clusters"]], "thoraxe.transcript_info.exon_clustering": [[4, "module-thoraxe.transcript_info.exon_clustering"]], "thoraxe.transcript_info.phases": [[4, "module-thoraxe.transcript_info.phases"]], "thoraxe.transcript_info.transcript_info": [[4, "module-thoraxe.transcript_info.transcript_info"]], "dictseq2fasta() (in module thoraxe.transcript_query.transcript_query)": [[5, "thoraxe.transcript_query.transcript_query.dictseq2fasta"]], "filter_ortho() (in module thoraxe.transcript_query.transcript_query)": [[5, "thoraxe.transcript_query.transcript_query.filter_ortho"]], "generic_ensembl_rest_request() (in module thoraxe.transcript_query.transcript_query)": [[5, "thoraxe.transcript_query.transcript_query.generic_ensembl_rest_request"]], "get_biomart_exons_annot() (in module thoraxe.transcript_query.transcript_query)": [[5, "thoraxe.transcript_query.transcript_query.get_biomart_exons_annot"]], "get_exons_sequences() (in module thoraxe.transcript_query.transcript_query)": [[5, "thoraxe.transcript_query.transcript_query.get_exons_sequences"]], "get_geneids_from_symbol() (in module thoraxe.transcript_query.transcript_query)": [[5, "thoraxe.transcript_query.transcript_query.get_geneids_from_symbol"]], "get_genetree() (in module thoraxe.transcript_query.transcript_query)": [[5, "thoraxe.transcript_query.transcript_query.get_genetree"]], "get_listofexons() (in module thoraxe.transcript_query.transcript_query)": [[5, "thoraxe.transcript_query.transcript_query.get_listofexons"]], "get_listoftranscripts() (in module thoraxe.transcript_query.transcript_query)": [[5, "thoraxe.transcript_query.transcript_query.get_listoftranscripts"]], "get_orthologs() (in module thoraxe.transcript_query.transcript_query)": [[5, "thoraxe.transcript_query.transcript_query.get_orthologs"]], "get_transcripts_orthologs() (in module thoraxe.transcript_query.transcript_query)": [[5, "thoraxe.transcript_query.transcript_query.get_transcripts_orthologs"]], "is_esemble_id() (in module thoraxe.transcript_query.transcript_query)": [[5, "thoraxe.transcript_query.transcript_query.is_esemble_id"]], "lodict2csv() (in module thoraxe.transcript_query.transcript_query)": [[5, "thoraxe.transcript_query.transcript_query.lodict2csv"]], "main() (in module thoraxe.transcript_query.transcript_query)": [[5, "thoraxe.transcript_query.transcript_query.main"]], "parse_command_line() (in module thoraxe.transcript_query.transcript_query)": [[5, "thoraxe.transcript_query.transcript_query.parse_command_line"]], "save_ensembl_version() (in module thoraxe.transcript_query.transcript_query)": [[5, "thoraxe.transcript_query.transcript_query.save_ensembl_version"]], "thoraxe.transcript_query": [[5, "module-thoraxe.transcript_query"]], "thoraxe.transcript_query.transcript_query": [[5, "module-thoraxe.transcript_query.transcript_query"]], "write_tsl_file() (in module thoraxe.transcript_query.transcript_query)": [[5, "thoraxe.transcript_query.transcript_query.write_tsl_file"]], "check_species_name() (in module thoraxe.utils.species)": [[6, "thoraxe.utils.species.check_species_name"]], "create_subfolder() (in module thoraxe.utils.folders)": [[6, "thoraxe.utils.folders.create_subfolder"]], "get_species_list() (in module thoraxe.utils.species)": [[6, "thoraxe.utils.species.get_species_list"]], "is_windows() (in module thoraxe.utils.windows)": [[6, "thoraxe.utils.windows.is_windows"]], "split_exons() (in module thoraxe.utils.split)": [[6, "thoraxe.utils.split.split_exons"]], "split_lengths() (in module thoraxe.utils.split)": [[6, "thoraxe.utils.split.split_lengths"]], "split_seqs() (in module thoraxe.utils.split)": [[6, "thoraxe.utils.split.split_seqs"]], "thoraxe.utils": [[6, "module-thoraxe.utils"]], "thoraxe.utils.folders": [[6, "module-thoraxe.utils.folders"]], "thoraxe.utils.species": [[6, "module-thoraxe.utils.species"]], "thoraxe.utils.split": [[6, "module-thoraxe.utils.split"]], "thoraxe.utils.windows": [[6, "module-thoraxe.utils.windows"]]}})