

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>thoraxe package &mdash; ThorAxe 0.1.0 documentation</title>
  

  
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />

  
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="thoraxe.add_transcripts package" href="thoraxe.add_transcripts.html" />
    <link rel="prev" title="thoraxe" href="modules.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../index.html" class="icon icon-home"> ThorAxe
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../introduction.html">Quick Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../programs/programs.html">Command line programs</a></li>
<li class="toctree-l1"><a class="reference internal" href="../output.html">Output description</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="modules.html">Module API</a><ul class="current">
<li class="toctree-l2 current"><a class="current reference internal" href="#">thoraxe package</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#subpackages">Subpackages</a><ul>
<li class="toctree-l4"><a class="reference internal" href="thoraxe.add_transcripts.html">thoraxe.add_transcripts package</a></li>
<li class="toctree-l4"><a class="reference internal" href="thoraxe.subexons.html">thoraxe.subexons package</a></li>
<li class="toctree-l4"><a class="reference internal" href="thoraxe.transcript_info.html">thoraxe.transcript_info package</a></li>
<li class="toctree-l4"><a class="reference internal" href="thoraxe.transcript_query.html">thoraxe.transcript_query package</a></li>
<li class="toctree-l4"><a class="reference internal" href="thoraxe.utils.html">thoraxe.utils package</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#submodules">Submodules</a></li>
<li class="toctree-l3"><a class="reference internal" href="#module-thoraxe.thoraxe">thoraxe.thoraxe module</a></li>
<li class="toctree-l3"><a class="reference internal" href="#module-thoraxe.version">thoraxe.version module</a></li>
<li class="toctree-l3"><a class="reference internal" href="#module-thoraxe">Module contents</a></li>
</ul>
</li>
</ul>
</li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">ThorAxe</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="modules.html">thoraxe</a> &raquo;</li>
        
      <li>thoraxe package</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="../_sources/api/thoraxe.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="thoraxe-package">
<h1>thoraxe package<a class="headerlink" href="#thoraxe-package" title="Permalink to this headline">¶</a></h1>
<div class="section" id="subpackages">
<h2>Subpackages<a class="headerlink" href="#subpackages" title="Permalink to this headline">¶</a></h2>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="thoraxe.add_transcripts.html">thoraxe.add_transcripts package</a><ul>
<li class="toctree-l2"><a class="reference internal" href="thoraxe.add_transcripts.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="thoraxe.add_transcripts.html#module-thoraxe.add_transcripts.add_transcripts">thoraxe.add_transcripts.add_transcripts module</a></li>
<li class="toctree-l2"><a class="reference internal" href="thoraxe.add_transcripts.html#module-thoraxe.add_transcripts">Module contents</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="thoraxe.subexons.html">thoraxe.subexons package</a><ul>
<li class="toctree-l2"><a class="reference internal" href="thoraxe.subexons.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="thoraxe.subexons.html#module-thoraxe.subexons.alignment">thoraxe.subexons.alignment module</a></li>
<li class="toctree-l2"><a class="reference internal" href="thoraxe.subexons.html#module-thoraxe.subexons.ases">thoraxe.subexons.ases module</a></li>
<li class="toctree-l2"><a class="reference internal" href="thoraxe.subexons.html#module-thoraxe.subexons.graph">thoraxe.subexons.graph module</a></li>
<li class="toctree-l2"><a class="reference internal" href="thoraxe.subexons.html#module-thoraxe.subexons.phylosofs">thoraxe.subexons.phylosofs module</a></li>
<li class="toctree-l2"><a class="reference internal" href="thoraxe.subexons.html#module-thoraxe.subexons.plot">thoraxe.subexons.plot module</a></li>
<li class="toctree-l2"><a class="reference internal" href="thoraxe.subexons.html#module-thoraxe.subexons.rescue">thoraxe.subexons.rescue module</a></li>
<li class="toctree-l2"><a class="reference internal" href="thoraxe.subexons.html#module-thoraxe.subexons.subexons">thoraxe.subexons.subexons module</a></li>
<li class="toctree-l2"><a class="reference internal" href="thoraxe.subexons.html#module-thoraxe.subexons.tidy">thoraxe.subexons.tidy module</a></li>
<li class="toctree-l2"><a class="reference internal" href="thoraxe.subexons.html#module-thoraxe.subexons">Module contents</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="thoraxe.transcript_info.html">thoraxe.transcript_info package</a><ul>
<li class="toctree-l2"><a class="reference internal" href="thoraxe.transcript_info.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="thoraxe.transcript_info.html#module-thoraxe.transcript_info.clusters">thoraxe.transcript_info.clusters module</a></li>
<li class="toctree-l2"><a class="reference internal" href="thoraxe.transcript_info.html#module-thoraxe.transcript_info.exon_clustering">thoraxe.transcript_info.exon_clustering module</a></li>
<li class="toctree-l2"><a class="reference internal" href="thoraxe.transcript_info.html#module-thoraxe.transcript_info.phases">thoraxe.transcript_info.phases module</a></li>
<li class="toctree-l2"><a class="reference internal" href="thoraxe.transcript_info.html#module-thoraxe.transcript_info.transcript_info">thoraxe.transcript_info.transcript_info module</a></li>
<li class="toctree-l2"><a class="reference internal" href="thoraxe.transcript_info.html#module-thoraxe.transcript_info">Module contents</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="thoraxe.transcript_query.html">thoraxe.transcript_query package</a><ul>
<li class="toctree-l2"><a class="reference internal" href="thoraxe.transcript_query.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="thoraxe.transcript_query.html#module-thoraxe.transcript_query.transcript_query">thoraxe.transcript_query.transcript_query module</a></li>
<li class="toctree-l2"><a class="reference internal" href="thoraxe.transcript_query.html#module-thoraxe.transcript_query">Module contents</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="thoraxe.utils.html">thoraxe.utils package</a><ul>
<li class="toctree-l2"><a class="reference internal" href="thoraxe.utils.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="thoraxe.utils.html#module-thoraxe.utils.folders">thoraxe.utils.folders module</a></li>
<li class="toctree-l2"><a class="reference internal" href="thoraxe.utils.html#module-thoraxe.utils.species">thoraxe.utils.species module</a></li>
<li class="toctree-l2"><a class="reference internal" href="thoraxe.utils.html#module-thoraxe.utils.split">thoraxe.utils.split module</a></li>
<li class="toctree-l2"><a class="reference internal" href="thoraxe.utils.html#module-thoraxe.utils">Module contents</a></li>
</ul>
</li>
</ul>
</div>
</div>
<div class="section" id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="module-thoraxe.thoraxe">
<span id="thoraxe-thoraxe-module"></span><h2>thoraxe.thoraxe module<a class="headerlink" href="#module-thoraxe.thoraxe" title="Permalink to this headline">¶</a></h2>
<p>ThorAxe pipeline and script functions.</p>
<dl class="py function">
<dt class="sig sig-object py" id="thoraxe.thoraxe.add_s_exon_phases_and_coordinates">
<span class="sig-prename descclassname"><span class="pre">thoraxe.thoraxe.</span></span><span class="sig-name descname"><span class="pre">add_s_exon_phases_and_coordinates</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">tbl</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#thoraxe.thoraxe.add_s_exon_phases_and_coordinates" title="Permalink to this definition">¶</a></dt>
<dd><p>Add s-exon genomic coordinates and phases to the tidy s-exon table.</p>
<p>For each s-exon, it add the <code class="docutils literal notranslate"><span class="pre">S_exon_Genomic_Sequence</span></code> and their genomic
coordinates, starting at <code class="docutils literal notranslate"><span class="pre">1</span></code> and defining the closed interval:
<code class="docutils literal notranslate"><span class="pre">[S_exon_CodingStart,</span> <span class="pre">S_exon_CodingEnd]</span></code>. Where <code class="docutils literal notranslate"><span class="pre">S_exon_CodingStart</span></code>
is greater or equal to <code class="docutils literal notranslate"><span class="pre">S_exon_CodingEnd</span></code> for genes in the reverse
<code class="docutils literal notranslate"><span class="pre">Strand</span></code>. It add also the <code class="docutils literal notranslate"><span class="pre">S_exon_StartPhase</span></code> and <code class="docutils literal notranslate"><span class="pre">S_exon_EndPhase</span></code>
for those intervals.</p>
<p>An s-exon, defined at the protein level, can be an entire sub-exon or a
part of it. In cases where an s-exon is identical to a sub-exon, the
genomic coordinates and the phases are the same. However, when a sub-exon
is split into several s-exons, coordinates and phases need to be calculated:</p>
<p><strong>Example 1: s-exons in the positive strand.</strong></p>
<p>The following sub-exon shares codons with the previous and following
sub-exon. In this example, <code class="docutils literal notranslate"><span class="pre">-</span></code> represents introns and lower case
characters represent exon nucleotides. The start and end phases are <code class="docutils literal notranslate"><span class="pre">1</span></code>
and <code class="docutils literal notranslate"><span class="pre">2</span></code> respectively:</p>
<p><code class="docutils literal notranslate"><span class="pre">...v----vvwwwxxxyyyzz-----z...</span></code></p>
<p>Because we are in the positive strand, the residue <code class="docutils literal notranslate"><span class="pre">V</span></code> coming from the
<code class="docutils literal notranslate"><span class="pre">vvv</span></code> shared codon is assigned to this sub-exon protein sequence. In the
same way, the <code class="docutils literal notranslate"><span class="pre">Z</span></code> residue coming from the <code class="docutils literal notranslate"><span class="pre">zzz</span></code> shared exon is assigned
to the beginning of the next sub-exon:</p>
<p><code class="docutils literal notranslate"><span class="pre">VWXY</span></code></p>
<p>Let’s say that this sub-exon is composed of two s-exons: <code class="docutils literal notranslate"><span class="pre">VW</span></code> and <code class="docutils literal notranslate"><span class="pre">XY</span></code>.
Then the corresponding genomic sequences are <code class="docutils literal notranslate"><span class="pre">vvwww</span></code> and <code class="docutils literal notranslate"><span class="pre">xxxyyyzz</span></code> and
the <code class="docutils literal notranslate"><span class="pre">(start,</span> <span class="pre">end)</span></code> phases are <code class="docutils literal notranslate"><span class="pre">(1,</span> <span class="pre">0)</span></code> and <code class="docutils literal notranslate"><span class="pre">(0,</span> <span class="pre">2)</span></code>.</p>
<p><em>Genomic coordinates.</em></p>
<p>Let’s say that the first <code class="docutils literal notranslate"><span class="pre">v</span></code> has the genomic coordinate <code class="docutils literal notranslate"><span class="pre">1</span></code>:</p>
<p><code class="docutils literal notranslate"><span class="pre">v</span> <span class="pre">----</span> <span class="pre">vvwww</span> <span class="pre">xxxyyyzz</span> <span class="pre">-----</span> <span class="pre">z</span></code>
``           1 11111111 12222 2``
<code class="docutils literal notranslate"><span class="pre">1</span> <span class="pre">2345</span> <span class="pre">67890</span> <span class="pre">12345678</span> <span class="pre">90123</span> <span class="pre">4</span></code></p>
<p>In this way, the first s-exon has the genomic coordinates <code class="docutils literal notranslate"><span class="pre">[6,</span> <span class="pre">10]</span></code> and
the second has the genomic coordinates <code class="docutils literal notranslate"><span class="pre">[11,</span> <span class="pre">18]</span></code>. The first s-exon has
<code class="docutils literal notranslate"><span class="pre">2</span></code> amino-acid residues, but <code class="docutils literal notranslate"><span class="pre">5</span></code> instead of <code class="docutils literal notranslate"><span class="pre">6</span></code> bases because the
first codon is shared between two s-exons. In particular, the
amino-acid residue <code class="docutils literal notranslate"><span class="pre">V</span></code> needs two bases in this s-exon to complete the
codon started in the previous s-exon.</p>
<p><strong>Example 2: s-exons in the negative strand.</strong></p>
<p><code class="docutils literal notranslate"><span class="pre">v</span> <span class="pre">----</span> <span class="pre">vvwww</span> <span class="pre">xxxyyyzz</span> <span class="pre">-----</span> <span class="pre">z</span></code>
<code class="docutils literal notranslate"><span class="pre">2</span> <span class="pre">2222</span> <span class="pre">11111</span> <span class="pre">11111</span>&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160;&#160; <span class="pre">``</span>
<span class="pre">``4</span> <span class="pre">3210</span> <span class="pre">98765</span> <span class="pre">43210987</span> <span class="pre">65432</span> <span class="pre">1</span></code></p>
<p>If the same sequence belongs to the negative strand, the sub-exon protein
sequence is going to contain <code class="docutils literal notranslate"><span class="pre">Z</span></code> but not <code class="docutils literal notranslate"><span class="pre">V</span></code>: <code class="docutils literal notranslate"><span class="pre">WXYZ</span></code>. Therefore,
possible s-exons are <code class="docutils literal notranslate"><span class="pre">W</span></code> and <code class="docutils literal notranslate"><span class="pre">XYZ</span></code> and the corresponding genomic
sequences will be <code class="docutils literal notranslate"><span class="pre">vvwww</span></code> and <code class="docutils literal notranslate"><span class="pre">xxxyyyzz</span></code> and the <code class="docutils literal notranslate"><span class="pre">(start,</span> <span class="pre">end)</span></code>
phases will be <code class="docutils literal notranslate"><span class="pre">(1,</span> <span class="pre">0)</span></code> and <code class="docutils literal notranslate"><span class="pre">(0,</span> <span class="pre">2)</span></code>.</p>
<p>In this way, the first s-exon has the genomic coordinates <code class="docutils literal notranslate"><span class="pre">[19,</span> <span class="pre">15]</span></code> and
the second has the genomic coordinates <code class="docutils literal notranslate"><span class="pre">[14,</span> <span class="pre">7]</span></code>. The first s-exon has `
<cite>1`</cite> amino-acid residue, but <code class="docutils literal notranslate"><span class="pre">5</span></code> instead of <code class="docutils literal notranslate"><span class="pre">3</span></code> bases because the first
codon is shared between two s-exons. In particular, the amino-acid residue
<code class="docutils literal notranslate"><span class="pre">V</span></code> needs two bases in this s-exon to complete the codon started in the
previous s-exon.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="thoraxe.thoraxe.create_chimeric_msa">
<span class="sig-prename descclassname"><span class="pre">thoraxe.thoraxe.</span></span><span class="sig-name descname"><span class="pre">create_chimeric_msa</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">output_folder</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subexon_table</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene2speciesname</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">connected_subexons</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">clusters</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cutoff</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">30.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_col_number</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">4</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">aligner</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'ProGraphMSA'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">padding</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'XXXXXXXXXX'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">species_list</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">keep_single_subexons</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#thoraxe.thoraxe.create_chimeric_msa" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a dict from cluster to cluster data.</p>
<p>For each cluster, there is a tuple with the subexon dataframe, the
chimeric sequences and the msa.</p>
<p>This function can take a clusters argument with the list of ‘Cluster’
identifiers to use. If that list is not given, all the positive ‘Cluster’
identifiers from subexon_table are used.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="thoraxe.thoraxe.get_s_exon_msas">
<span class="sig-prename descclassname"><span class="pre">thoraxe.thoraxe.</span></span><span class="sig-name descname"><span class="pre">get_s_exon_msas</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">output_folder</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#thoraxe.thoraxe.get_s_exon_msas" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a dict of the s_exon MSAs.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="thoraxe.thoraxe.get_s_exons">
<span class="sig-prename descclassname"><span class="pre">thoraxe.thoraxe.</span></span><span class="sig-name descname"><span class="pre">get_s_exons</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">output_folder</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subexon_table</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene2speciesname</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">connected_subexons</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">minimum_len</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">4</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">coverage_cutoff</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">80.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">percent_identity_cutoff</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">30.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gap_open_penalty</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gap_extend_penalty</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">aligner</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'ProGraphMSA'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">padding</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'XXXXXXXXXX'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">movements</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">disintegration</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">species_list</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">keep_single_subexons</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#thoraxe.thoraxe.get_s_exons" title="Permalink to this definition">¶</a></dt>
<dd><p>Perform almost all the ThorAxe pipeline.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="thoraxe.thoraxe.get_subexons">
<span class="sig-prename descclassname"><span class="pre">thoraxe.thoraxe.</span></span><span class="sig-name descname"><span class="pre">get_subexons</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">transcript_table</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">minimum_len</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">4</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">coverage_cutoff</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">80.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">percent_identity_cutoff</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">30.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gap_open_penalty</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gap_extend_penalty</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#thoraxe.thoraxe.get_subexons" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a DataFrame with subexon information and clustered exons.</p>
<p>Exons are clustered according to their protein sequence using the blosum50
matrix and the following arguments: minimum_len, coverage_cutoff,
percent_identity_cutoff, gap_open_penalty and gap_extend_penalty.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="thoraxe.thoraxe.get_transcripts">
<span class="sig-prename descclassname"><span class="pre">thoraxe.thoraxe.</span></span><span class="sig-name descname"><span class="pre">get_transcripts</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input_folder</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_tsl_level</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">3.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">species_list</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#thoraxe.thoraxe.get_transcripts" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a DataFrame with the transcript information.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="thoraxe.thoraxe.main">
<span class="sig-prename descclassname"><span class="pre">thoraxe.thoraxe.</span></span><span class="sig-name descname"><span class="pre">main</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#thoraxe.thoraxe.main" title="Permalink to this definition">¶</a></dt>
<dd><p>Perform Pipeline.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="thoraxe.thoraxe.merge_clusters">
<span class="sig-prename descclassname"><span class="pre">thoraxe.thoraxe.</span></span><span class="sig-name descname"><span class="pre">merge_clusters</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">subexon_table</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#thoraxe.thoraxe.merge_clusters" title="Permalink to this definition">¶</a></dt>
<dd><p>Merge ‘Cluster’s that share subexons.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="thoraxe.thoraxe.parse_command_line">
<span class="sig-prename descclassname"><span class="pre">thoraxe.thoraxe.</span></span><span class="sig-name descname"><span class="pre">parse_command_line</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#thoraxe.thoraxe.parse_command_line" title="Permalink to this definition">¶</a></dt>
<dd><p>Parse command line.</p>
<p>It uses argparse to parse thoraxe’ command line arguments and returns
the argparse parser.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="thoraxe.thoraxe.update_subexon_table">
<span class="sig-prename descclassname"><span class="pre">thoraxe.thoraxe.</span></span><span class="sig-name descname"><span class="pre">update_subexon_table</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">subexon_table</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cluster2data</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#thoraxe.thoraxe.update_subexon_table" title="Permalink to this definition">¶</a></dt>
<dd><p>Update the subexon table by adding the s-exon information.</p>
</dd></dl>

</div>
<div class="section" id="module-thoraxe.version">
<span id="thoraxe-version-module"></span><h2>thoraxe.version module<a class="headerlink" href="#module-thoraxe.version" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="module-thoraxe">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-thoraxe" title="Permalink to this headline">¶</a></h2>
<p>thoraxe: Pipeline to disentangle homology relationships between exons.</p>
</div>
</div>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="thoraxe.add_transcripts.html" class="btn btn-neutral float-right" title="thoraxe.add_transcripts package" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
        <a href="modules.html" class="btn btn-neutral float-left" title="thoraxe" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2019, Diego Javier Zea, Hugues Richard &amp; Elodie Laine.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>