<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>thoraxe.subexons package &mdash; ThorAxe 0.1.0 documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/sphinx_highlight.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="thoraxe.transcript_info package" href="thoraxe.transcript_info.html" />
    <link rel="prev" title="thoraxe.add_transcripts package" href="thoraxe.add_transcripts.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="../index.html" class="icon icon-home"> ThorAxe
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../introduction.html">Quick Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../programs/programs.html">Command line programs</a></li>
<li class="toctree-l1"><a class="reference internal" href="../output.html">Output description</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="modules.html">Module API</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="thoraxe.html">thoraxe package</a><ul class="current">
<li class="toctree-l3 current"><a class="reference internal" href="thoraxe.html#subpackages">Subpackages</a><ul class="current">
<li class="toctree-l4"><a class="reference internal" href="thoraxe.add_transcripts.html">thoraxe.add_transcripts package</a></li>
<li class="toctree-l4 current"><a class="current reference internal" href="#">thoraxe.subexons package</a></li>
<li class="toctree-l4"><a class="reference internal" href="thoraxe.transcript_info.html">thoraxe.transcript_info package</a></li>
<li class="toctree-l4"><a class="reference internal" href="thoraxe.transcript_query.html">thoraxe.transcript_query package</a></li>
<li class="toctree-l4"><a class="reference internal" href="thoraxe.utils.html">thoraxe.utils package</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="thoraxe.html#submodules">Submodules</a></li>
<li class="toctree-l3"><a class="reference internal" href="thoraxe.html#module-thoraxe.thoraxe">thoraxe.thoraxe module</a></li>
<li class="toctree-l3"><a class="reference internal" href="thoraxe.html#module-thoraxe.version">thoraxe.version module</a></li>
<li class="toctree-l3"><a class="reference internal" href="thoraxe.html#module-thoraxe">Module contents</a></li>
</ul>
</li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">ThorAxe</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home"></a></li>
          <li class="breadcrumb-item"><a href="modules.html">thoraxe</a></li>
          <li class="breadcrumb-item"><a href="thoraxe.html">thoraxe package</a></li>
      <li class="breadcrumb-item active">thoraxe.subexons package</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/api/thoraxe.subexons.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="thoraxe-subexons-package">
<h1>thoraxe.subexons package<a class="headerlink" href="#thoraxe-subexons-package" title="Permalink to this heading"></a></h1>
<section id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Permalink to this heading"></a></h2>
</section>
<section id="module-thoraxe.subexons.alignment">
<span id="thoraxe-subexons-alignment-module"></span><h2>thoraxe.subexons.alignment module<a class="headerlink" href="#module-thoraxe.subexons.alignment" title="Permalink to this heading"></a></h2>
<p>alignment: Module to create the subexon MSA with ProGraphMSA.</p>
<dl class="py class">
<dt class="sig sig-object py" id="thoraxe.subexons.alignment.ColCluster">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">thoraxe.subexons.alignment.</span></span><span class="sig-name descname"><span class="pre">ColCluster</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">patterns</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">consensus</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">end</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#thoraxe.subexons.alignment.ColCluster" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="thoraxe.subexons.alignment.ColPattern">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">thoraxe.subexons.alignment.</span></span><span class="sig-name descname"><span class="pre">ColPattern</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pattern</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">end</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#thoraxe.subexons.alignment.ColPattern" title="Permalink to this definition"></a></dt>
<dd><dl class="py property">
<dt class="sig sig-object py" id="thoraxe.subexons.alignment.ColPattern.end">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">end</span></span><a class="headerlink" href="#thoraxe.subexons.alignment.ColPattern.end" title="Permalink to this definition"></a></dt>
<dd><p>Alias for field number 2</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="thoraxe.subexons.alignment.ColPattern.pattern">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">pattern</span></span><a class="headerlink" href="#thoraxe.subexons.alignment.ColPattern.pattern" title="Permalink to this definition"></a></dt>
<dd><p>Alias for field number 0</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="thoraxe.subexons.alignment.ColPattern.start">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">start</span></span><a class="headerlink" href="#thoraxe.subexons.alignment.ColPattern.start" title="Permalink to this definition"></a></dt>
<dd><p>Alias for field number 1</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="thoraxe.subexons.alignment.ProblematicSubexonBlock">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">thoraxe.subexons.alignment.</span></span><span class="sig-name descname"><span class="pre">ProblematicSubexonBlock</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sequence_index</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subexon</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subexon_blocks</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">block_type</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subexon_block_start</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subexon_block_end</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gap_block_start</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gap_block_end</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#thoraxe.subexons.alignment.ProblematicSubexonBlock" title="Permalink to this definition"></a></dt>
<dd><dl class="py property">
<dt class="sig sig-object py" id="thoraxe.subexons.alignment.ProblematicSubexonBlock.block_type">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">block_type</span></span><a class="headerlink" href="#thoraxe.subexons.alignment.ProblematicSubexonBlock.block_type" title="Permalink to this definition"></a></dt>
<dd><p>Alias for field number 3</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="thoraxe.subexons.alignment.ProblematicSubexonBlock.gap_block_end">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">gap_block_end</span></span><a class="headerlink" href="#thoraxe.subexons.alignment.ProblematicSubexonBlock.gap_block_end" title="Permalink to this definition"></a></dt>
<dd><p>Alias for field number 7</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="thoraxe.subexons.alignment.ProblematicSubexonBlock.gap_block_start">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">gap_block_start</span></span><a class="headerlink" href="#thoraxe.subexons.alignment.ProblematicSubexonBlock.gap_block_start" title="Permalink to this definition"></a></dt>
<dd><p>Alias for field number 6</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="thoraxe.subexons.alignment.ProblematicSubexonBlock.sequence_index">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">sequence_index</span></span><a class="headerlink" href="#thoraxe.subexons.alignment.ProblematicSubexonBlock.sequence_index" title="Permalink to this definition"></a></dt>
<dd><p>Alias for field number 0</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="thoraxe.subexons.alignment.ProblematicSubexonBlock.subexon">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">subexon</span></span><a class="headerlink" href="#thoraxe.subexons.alignment.ProblematicSubexonBlock.subexon" title="Permalink to this definition"></a></dt>
<dd><p>Alias for field number 1</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="thoraxe.subexons.alignment.ProblematicSubexonBlock.subexon_block_end">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">subexon_block_end</span></span><a class="headerlink" href="#thoraxe.subexons.alignment.ProblematicSubexonBlock.subexon_block_end" title="Permalink to this definition"></a></dt>
<dd><p>Alias for field number 5</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="thoraxe.subexons.alignment.ProblematicSubexonBlock.subexon_block_start">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">subexon_block_start</span></span><a class="headerlink" href="#thoraxe.subexons.alignment.ProblematicSubexonBlock.subexon_block_start" title="Permalink to this definition"></a></dt>
<dd><p>Alias for field number 4</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="thoraxe.subexons.alignment.ProblematicSubexonBlock.subexon_blocks">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">subexon_blocks</span></span><a class="headerlink" href="#thoraxe.subexons.alignment.ProblematicSubexonBlock.subexon_blocks" title="Permalink to this definition"></a></dt>
<dd><p>Alias for field number 2</p>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="thoraxe.subexons.alignment.cluster_subexon_blocks">
<span class="sig-prename descclassname"><span class="pre">thoraxe.subexons.alignment.</span></span><span class="sig-name descname"><span class="pre">cluster_subexon_blocks</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">blocks</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">margin</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#thoraxe.subexons.alignment.cluster_subexon_blocks" title="Permalink to this definition"></a></dt>
<dd><p>Cluster sub-exon blocks starting or ending in the same column.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="thoraxe.subexons.alignment.column_clusters">
<span class="sig-prename descclassname"><span class="pre">thoraxe.subexons.alignment.</span></span><span class="sig-name descname"><span class="pre">column_clusters</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">colpatterns</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#thoraxe.subexons.alignment.column_clusters" title="Permalink to this definition"></a></dt>
<dd><p>Return a ColCluster list from a ColPattern list.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="thoraxe.subexons.alignment.column_patterns">
<span class="sig-prename descclassname"><span class="pre">thoraxe.subexons.alignment.</span></span><span class="sig-name descname"><span class="pre">column_patterns</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">msa_matrix</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#thoraxe.subexons.alignment.column_patterns" title="Permalink to this definition"></a></dt>
<dd><p>Return a ColPattern list from the msa_matrix.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="thoraxe.subexons.alignment.create_chimeric_sequences">
<span class="sig-prename descclassname"><span class="pre">thoraxe.subexons.alignment.</span></span><span class="sig-name descname"><span class="pre">create_chimeric_sequences</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">subexon_table</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subexon_matrix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">connected_subexons</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">padding</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'XXXXXXXXXX'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#thoraxe.subexons.alignment.create_chimeric_sequences" title="Permalink to this definition"></a></dt>
<dd><p>Create chimeric sequence for ProGraphMSA.</p>
<p>It returns a Dict from ‘GeneID’ to a tuple with the chimeric
sequence and a Dict from ‘SubexonIndex’ to …</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="thoraxe.subexons.alignment.create_msa_matrix">
<span class="sig-prename descclassname"><span class="pre">thoraxe.subexons.alignment.</span></span><span class="sig-name descname"><span class="pre">create_msa_matrix</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">chimerics</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">msa</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#thoraxe.subexons.alignment.create_msa_matrix" title="Permalink to this definition"></a></dt>
<dd><p>Convert a msa from chimerics to a matrix.</p>
<p>Each cell has the subexon number (Index) or nan for gaps and padding.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="thoraxe.subexons.alignment.create_subexon_matrix">
<span class="sig-prename descclassname"><span class="pre">thoraxe.subexons.alignment.</span></span><span class="sig-name descname"><span class="pre">create_subexon_matrix</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">subexon_table</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#thoraxe.subexons.alignment.create_subexon_matrix" title="Permalink to this definition"></a></dt>
<dd><p>Return a binary matrix showing subexon presence in transcripts.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="thoraxe.subexons.alignment.delete_padding">
<span class="sig-prename descclassname"><span class="pre">thoraxe.subexons.alignment.</span></span><span class="sig-name descname"><span class="pre">delete_padding</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">padding</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#thoraxe.subexons.alignment.delete_padding" title="Permalink to this definition"></a></dt>
<dd><p>Replace padding by gaps.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">delete_padding</span><span class="p">(</span><span class="s2">&quot;MHGL--XXXX-XXX--XXXKLMHSXXXXXXX-X-X-X&quot;</span><span class="p">,</span> <span class="s2">&quot;XXXXXXXXXX&quot;</span><span class="p">)</span>
<span class="go">&#39;MHGL---------------KLMHS-------------&#39;</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">delete_padding</span><span class="p">(</span><span class="s2">&quot;MHGLXKLMHSX&quot;</span><span class="p">,</span> <span class="s2">&quot;XXXXXXXXXX&quot;</span><span class="p">)</span>
<span class="go">&#39;MHGLXKLMHSX&#39;</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="thoraxe.subexons.alignment.delete_subexons">
<span class="sig-prename descclassname"><span class="pre">thoraxe.subexons.alignment.</span></span><span class="sig-name descname"><span class="pre">delete_subexons</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">subexon_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chimerics</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">msa</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cutoff</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">30.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_col_number</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">4</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">keep_single_subexons</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#thoraxe.subexons.alignment.delete_subexons" title="Permalink to this definition"></a></dt>
<dd><p>Return the list of ‘SubexonIDCluster’ to delete from ‘Cluster’.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="thoraxe.subexons.alignment.disintegration">
<span class="sig-prename descclassname"><span class="pre">thoraxe.subexons.alignment.</span></span><span class="sig-name descname"><span class="pre">disintegration</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">msa_numpy</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">msa_matrix</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#thoraxe.subexons.alignment.disintegration" title="Permalink to this definition"></a></dt>
<dd><p>Disintegrate a one-residue length s-exons.</p>
<p>It merges their residues to the contiguous s-exons.
Return the modified MSA and sub-exon matrix.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="thoraxe.subexons.alignment.gene2species">
<span class="sig-prename descclassname"><span class="pre">thoraxe.subexons.alignment.</span></span><span class="sig-name descname"><span class="pre">gene2species</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">transcript_data</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#thoraxe.subexons.alignment.gene2species" title="Permalink to this definition"></a></dt>
<dd><p>Return the a dict from ‘GeneID’ to ‘Species’.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="thoraxe.subexons.alignment.get_consensus">
<span class="sig-prename descclassname"><span class="pre">thoraxe.subexons.alignment.</span></span><span class="sig-name descname"><span class="pre">get_consensus</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">msa</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">threshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.5</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#thoraxe.subexons.alignment.get_consensus" title="Permalink to this definition"></a></dt>
<dd><p>Return the consensus sequence of the MSA.</p>
<p>It returns the amino-acid residue that appears in more than <cite>threshold</cite>
fraction of the sequences without counting sequences that has gaps in
that column. If any residue is present in more sequences than the
<cite>threshold</cite>, it introduces an <cite>X</cite> to indicate the ambiguity.
If there are more than 50% gaps in a column, it displays the residue
in lowercase.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Alphabet</span> <span class="kn">import</span> <span class="n">IUPAC</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.SeqRecord</span> <span class="kn">import</span> <span class="n">SeqRecord</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">Bio.Align</span> <span class="kn">import</span> <span class="n">MultipleSeqAlignment</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq_1</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;SEEEEACCC&quot;</span><span class="p">,</span> <span class="n">IUPAC</span><span class="o">.</span><span class="n">protein</span><span class="p">),</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;I&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq_2</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;SSEEEGCC-&quot;</span><span class="p">,</span> <span class="n">IUPAC</span><span class="o">.</span><span class="n">protein</span><span class="p">),</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;II&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq_3</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;SSSEEKC--&quot;</span><span class="p">,</span> <span class="n">IUPAC</span><span class="o">.</span><span class="n">protein</span><span class="p">),</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;III&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq_4</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;SSSSEH---&quot;</span><span class="p">,</span> <span class="n">IUPAC</span><span class="o">.</span><span class="n">protein</span><span class="p">),</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;IIII&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">msa</span> <span class="o">=</span> <span class="n">MultipleSeqAlignment</span><span class="p">([</span><span class="n">seq_1</span><span class="p">,</span> <span class="n">seq_2</span><span class="p">,</span> <span class="n">seq_3</span><span class="p">,</span> <span class="n">seq_4</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">get_consensus</span><span class="p">(</span><span class="n">msa</span><span class="p">)</span>
<span class="go">&#39;SSXEEXCCc&#39;</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="thoraxe.subexons.alignment.get_gene_ids">
<span class="sig-prename descclassname"><span class="pre">thoraxe.subexons.alignment.</span></span><span class="sig-name descname"><span class="pre">get_gene_ids</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">msa</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#thoraxe.subexons.alignment.get_gene_ids" title="Permalink to this definition"></a></dt>
<dd><p>Take a biopython MSA and return a list with the gene (sequence) ids.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="thoraxe.subexons.alignment.get_subexon_boundaries">
<span class="sig-prename descclassname"><span class="pre">thoraxe.subexons.alignment.</span></span><span class="sig-name descname"><span class="pre">get_subexon_boundaries</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">col_clusters</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#thoraxe.subexons.alignment.get_subexon_boundaries" title="Permalink to this definition"></a></dt>
<dd><p>Returns a dict from (sequence number, subexon id) to start and end columns.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="thoraxe.subexons.alignment.get_submsa">
<span class="sig-prename descclassname"><span class="pre">thoraxe.subexons.alignment.</span></span><span class="sig-name descname"><span class="pre">get_submsa</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">msa</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seq_ids</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#thoraxe.subexons.alignment.get_submsa" title="Permalink to this definition"></a></dt>
<dd><p>Return the sub-MSA with <cite>seq_ids</cite>, deleting full gap columns.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="thoraxe.subexons.alignment.impute_missing_s_exon">
<span class="sig-prename descclassname"><span class="pre">thoraxe.subexons.alignment.</span></span><span class="sig-name descname"><span class="pre">impute_missing_s_exon</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">table</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">column</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'S_exons'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key_columns</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#thoraxe.subexons.alignment.impute_missing_s_exon" title="Permalink to this definition"></a></dt>
<dd><p>Replace column values that do not conform to the naming by 0_number.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="thoraxe.subexons.alignment.move_block">
<span class="sig-prename descclassname"><span class="pre">thoraxe.subexons.alignment.</span></span><span class="sig-name descname"><span class="pre">move_block</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">matrix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">row</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">block_start</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">block_end</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">destination_start</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">destination_end</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#thoraxe.subexons.alignment.move_block" title="Permalink to this definition"></a></dt>
<dd><p>Move in-place a block in the (numpy) matrix.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">mat</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">matrix</span><span class="p">([[</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">],</span> <span class="p">[</span><span class="mi">5</span><span class="p">,</span><span class="mi">6</span><span class="p">,</span><span class="mi">7</span><span class="p">,</span><span class="mi">8</span><span class="p">,</span><span class="mi">9</span><span class="p">]])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">move_block</span><span class="p">(</span><span class="n">mat</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">5</span><span class="p">)</span>
<span class="go">matrix([[3, 4, 2, 0, 1],</span>
<span class="go">        [5, 6, 7, 8, 9]])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">move_block</span><span class="p">(</span><span class="n">mat</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">5</span><span class="p">)</span>
<span class="go">matrix([[3, 4, 2, 0, 1],</span>
<span class="go">        [8, 9, 7, 5, 6]])</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="thoraxe.subexons.alignment.move_problematic_block_clusters">
<span class="sig-prename descclassname"><span class="pre">thoraxe.subexons.alignment.</span></span><span class="sig-name descname"><span class="pre">move_problematic_block_clusters</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">msa</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">msa_matrix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_res_block</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_gap_block</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#thoraxe.subexons.alignment.move_problematic_block_clusters" title="Permalink to this definition"></a></dt>
<dd><p>Find and move all the problematic block clusters in a chimeric MSA.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="thoraxe.subexons.alignment.move_subexon_block">
<span class="sig-prename descclassname"><span class="pre">thoraxe.subexons.alignment.</span></span><span class="sig-name descname"><span class="pre">move_subexon_block</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">msa_numpy</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">msa_matrix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subexon_block</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#thoraxe.subexons.alignment.move_subexon_block" title="Permalink to this definition"></a></dt>
<dd><p>Swap a sub-exon block within its gap block.</p>
<p>The movement depends on the problematic sub-exon block position, e.g.:</p>
<p><a href="#id1"><span class="problematic" id="id2">``</span></a>`
from: [1.0, 1.0, 1.0, nan, nan, nan, 1.0, 2.0, 2.0, 2.0] # last_block
to:   [1.0, 1.0, 1.0, 1.0, nan, nan, nan, 2.0, 2.0, 2.0]</p>
<p>from: [1.0, 1.0, 1.0, 2.0, nan, nan, nan, 2.0, 2.0, 2.0] # first_block
to:   [1.0, 1.0, 1.0, nan, nan, nan, 2.0, 2.0, 2.0, 2.0]
<a href="#id3"><span class="problematic" id="id4">``</span></a><a href="#id5"><span class="problematic" id="id6">`</span></a></p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="thoraxe.subexons.alignment.msa2sequences">
<span class="sig-prename descclassname"><span class="pre">thoraxe.subexons.alignment.</span></span><span class="sig-name descname"><span class="pre">msa2sequences</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">msa</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_ids</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">padding</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#thoraxe.subexons.alignment.msa2sequences" title="Permalink to this definition"></a></dt>
<dd><p>Return str sequences from msa.</p>
<p>It also checks gene_ids and replaces padding by gaps.
Full gap columns are deleted.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="thoraxe.subexons.alignment.msa_matrices">
<span class="sig-prename descclassname"><span class="pre">thoraxe.subexons.alignment.</span></span><span class="sig-name descname"><span class="pre">msa_matrices</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">subexon_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chimerics</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">msa</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#thoraxe.subexons.alignment.msa_matrices" title="Permalink to this definition"></a></dt>
<dd><p>Return msa (residues) and subexon id cluster matrices from msa.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="thoraxe.subexons.alignment.problematic_block_clusters">
<span class="sig-prename descclassname"><span class="pre">thoraxe.subexons.alignment.</span></span><span class="sig-name descname"><span class="pre">problematic_block_clusters</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">msa_matrix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_res_block</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_gap_block</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#thoraxe.subexons.alignment.problematic_block_clusters" title="Permalink to this definition"></a></dt>
<dd><p>Return a Dict of problematic sub-exon block clusters.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="thoraxe.subexons.alignment.problematic_subexon_blocks">
<span class="sig-prename descclassname"><span class="pre">thoraxe.subexons.alignment.</span></span><span class="sig-name descname"><span class="pre">problematic_subexon_blocks</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">subexons</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">starts</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ends</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sequence</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_res_block</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_gap_block</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#thoraxe.subexons.alignment.problematic_subexon_blocks" title="Permalink to this definition"></a></dt>
<dd><p>Return a list of ProblematicSubexonBlocks.</p>
<p>In particular, a sub-exon block is problematic if it has &lt;= max_res_block
and it is separeated from the rest of the sub-exon by a gap block
=&gt; min_gap_block.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">problematic_subexon_blocks</span><span class="p">([</span><span class="mf">1.0</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">,</span> <span class="mf">1.0</span><span class="p">,</span> <span class="mf">2.0</span><span class="p">],</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">9</span><span class="p">,</span> <span class="mi">10</span><span class="p">],</span> <span class="p">[</span><span class="mi">3</span><span class="p">,</span> <span class="mi">9</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="mi">13</span><span class="p">],</span> <span class="mi">0</span><span class="p">)</span>
<span class="go">[ProblematicSubexonBlock(sequence_index=0, subexon=1.0, subexon_blocks=2, block_type=&#39;last_block&#39;, subexon_block_start=9, subexon_block_end=10, gap_block_start=3, gap_block_end=9)]</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="thoraxe.subexons.alignment.read_msa_fasta">
<span class="sig-prename descclassname"><span class="pre">thoraxe.subexons.alignment.</span></span><span class="sig-name descname"><span class="pre">read_msa_fasta</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">msa_file</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#thoraxe.subexons.alignment.read_msa_fasta" title="Permalink to this definition"></a></dt>
<dd><p>Return a BioPython’s alignment object from the fasta msa_file.</p>
<p>Return None if the file is empty.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="thoraxe.subexons.alignment.resume_seq">
<span class="sig-prename descclassname"><span class="pre">thoraxe.subexons.alignment.</span></span><span class="sig-name descname"><span class="pre">resume_seq</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">matrix_row</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#thoraxe.subexons.alignment.resume_seq" title="Permalink to this definition"></a></dt>
<dd><p>Return the sub-exon and gap blocks in the msa matrix.</p>
<p>The function return three lists, a list of sub-exon ids and NaNs, i.e.
the block id, a list with the column indices where the block starts and a
list with the block ends.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">resume_seq</span><span class="p">([</span><span class="mf">1.0</span><span class="p">,</span> <span class="mf">1.0</span><span class="p">,</span> <span class="mf">1.0</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">,</span> <span class="mf">1.0</span><span class="p">,</span> <span class="mf">2.0</span><span class="p">,</span> <span class="mf">2.0</span><span class="p">,</span> <span class="mf">2.0</span><span class="p">])</span>
<span class="go">([1.0, nan, 1.0, 2.0], [0, 3, 6, 7], [3, 6, 7, 10])</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="thoraxe.subexons.alignment.run_aligner">
<span class="sig-prename descclassname"><span class="pre">thoraxe.subexons.alignment.</span></span><span class="sig-name descname"><span class="pre">run_aligner</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">chimerics</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'alignment.fasta'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">aligner</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'ProGraphMSA'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#thoraxe.subexons.alignment.run_aligner" title="Permalink to this definition"></a></dt>
<dd><p>Run ProGraphMSA in the chimeric sequences and return the output file.</p>
<p>You can pass arguments using aligner (default: ‘ProGraphMSA’), e.g:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">aligner</span><span class="o">=</span><span class="s1">&#39;ProGraphMSA --mldist_gap&#39;</span>
</pre></div>
</div>
<p>You need ProGraphMSA installed to run this function. You can install it
from: <a class="reference external" href="https://github.com/acg-team/ProGraphMSA">https://github.com/acg-team/ProGraphMSA</a></p>
<p>If you are using Windows 10 and you have installed ProGraphMSA in
Ubuntu using the ‘Windows Subsystem for Linux’, you can try with the
following options:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">aligner</span><span class="o">=</span><span class="s1">&#39;ubuntu.exe -c ProGraphMSA&#39;</span>
<span class="n">aligner</span><span class="o">=</span><span class="s1">&#39;bash.exe -c ProGraphMSA&#39;</span>
<span class="n">aligner</span><span class="o">=</span><span class="s1">&#39;wsl.exe ProGraphMSA&#39;</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="thoraxe.subexons.alignment.save_s_exons">
<span class="sig-prename descclassname"><span class="pre">thoraxe.subexons.alignment.</span></span><span class="sig-name descname"><span class="pre">save_s_exons</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">subexon_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sequences</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_ids</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">colclusters</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_folder</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#thoraxe.subexons.alignment.save_s_exons" title="Permalink to this definition"></a></dt>
<dd><p>It saves the information about s-exons.</p>
<p>It takes a list of sequences, like the one returned by msa2sequences.
Return subexon_df with the orthologous exonic region (s-exon) information.
For each s-exon saves a fasta MSA in the output_folder.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="thoraxe.subexons.alignment.score_solution">
<span class="sig-prename descclassname"><span class="pre">thoraxe.subexons.alignment.</span></span><span class="sig-name descname"><span class="pre">score_solution</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">msa_matrix</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#thoraxe.subexons.alignment.score_solution" title="Permalink to this definition"></a></dt>
<dd><p>Return the number of times a sub-exon cross an s-exon boundary.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="thoraxe.subexons.alignment.sort_species">
<span class="sig-prename descclassname"><span class="pre">thoraxe.subexons.alignment.</span></span><span class="sig-name descname"><span class="pre">sort_species</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">chimerics</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene2sp</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">species_list</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#thoraxe.subexons.alignment.sort_species" title="Permalink to this definition"></a></dt>
<dd><p>Sort chimerics using the gene2species output and the species_list.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="thoraxe.subexons.alignment.subexon_connectivity">
<span class="sig-prename descclassname"><span class="pre">thoraxe.subexons.alignment.</span></span><span class="sig-name descname"><span class="pre">subexon_connectivity</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">subexon_table</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id_column</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'SubexonIDCluster'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#thoraxe.subexons.alignment.subexon_connectivity" title="Permalink to this definition"></a></dt>
<dd><p>Return a set of connected subexon tuples.</p>
<p>Each tuple has two elements, ‘SubexonIDCluster’ (directed) pairs.
The first subexon appears before of the second in at least one transcript.</p>
</dd></dl>

</section>
<section id="module-thoraxe.subexons.ases">
<span id="thoraxe-subexons-ases-module"></span><h2>thoraxe.subexons.ases module<a class="headerlink" href="#module-thoraxe.subexons.ases" title="Permalink to this heading"></a></h2>
<p>ases: Function to detect conserved alternative splicing events.</p>
<p>It finds the canonical path in the splice graph to detect conserved ASEs.</p>
<dl class="py class">
<dt class="sig sig-object py" id="thoraxe.subexons.ases.PathData">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">thoraxe.subexons.ases.</span></span><span class="sig-name descname"><span class="pre">PathData</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">Genes</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">Transcripts</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#thoraxe.subexons.ases.PathData" title="Permalink to this definition"></a></dt>
<dd><dl class="py property">
<dt class="sig sig-object py" id="thoraxe.subexons.ases.PathData.Genes">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">Genes</span></span><a class="headerlink" href="#thoraxe.subexons.ases.PathData.Genes" title="Permalink to this definition"></a></dt>
<dd><p>Alias for field number 0</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="thoraxe.subexons.ases.PathData.Transcripts">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">Transcripts</span></span><a class="headerlink" href="#thoraxe.subexons.ases.PathData.Transcripts" title="Permalink to this definition"></a></dt>
<dd><p>Alias for field number 1</p>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="thoraxe.subexons.ases.conserved_ases">
<span class="sig-prename descclassname"><span class="pre">thoraxe.subexons.ases.</span></span><span class="sig-name descname"><span class="pre">conserved_ases</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">s_exon_table</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">transcript_table</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">s_exon_msas</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">graph_file_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_genes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_transcripts</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">column_order</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">delim</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'/'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#thoraxe.subexons.ases.conserved_ases" title="Permalink to this definition"></a></dt>
<dd><p>Return two DataFrames.</p>
<ul class="simple">
<li><p>A table of transcripts/paths.</p></li>
<li><p>A table of the conserved alternative splicing events detected.</p></li>
</ul>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="thoraxe.subexons.ases.create_ases_table">
<span class="sig-prename descclassname"><span class="pre">thoraxe.subexons.ases.</span></span><span class="sig-name descname"><span class="pre">create_ases_table</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">s_exon_msas</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">s_exon_table</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">events</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">delim</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'/'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#thoraxe.subexons.ases.create_ases_table" title="Permalink to this definition"></a></dt>
<dd><p>Create an ASE data frame from the dictionary returned by <cite>detect_ases</cite>.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="thoraxe.subexons.ases.detect_ases">
<span class="sig-prename descclassname"><span class="pre">thoraxe.subexons.ases.</span></span><span class="sig-name descname"><span class="pre">detect_ases</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">path_table</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_genes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_transcripts</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">delim</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'/'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#thoraxe.subexons.ases.detect_ases" title="Permalink to this definition"></a></dt>
<dd><p>Return a dictionary from ASEs to their data.</p>
<p>It takes as imput a pandas DataFrame object with the path table.
Each key of the output dictionary is a tuple
(canonical_path, alternative_path) and the value is a tuple of:</p>
<blockquote>
<div><ul class="simple">
<li><p>one PathData tuple for the canonical path</p></li>
<li><p>and one for the alternative path;</p></li>
<li><p>the transcript weighted conservation for the canonical sub-path,</p></li>
<li><p>and for the alternative sub-path,</p></li>
<li><p>the type of alternative splicing event, e.g. “insertion”;</p></li>
<li><p>the mutually exclusivity of the event (”” for non mutually exclusive);</p></li>
<li><p>one list of mutually exclusive s-exons in canonical paths</p></li>
<li><p>and one for the alternative paths.</p></li>
</ul>
</div></blockquote>
<p>The last two list are paired, i.e. the two first s-exons are mutually
exclusive.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="thoraxe.subexons.ases.get_transcript_scores">
<span class="sig-prename descclassname"><span class="pre">thoraxe.subexons.ases.</span></span><span class="sig-name descname"><span class="pre">get_transcript_scores</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">s_exon_table</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">transcript_table</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">graph</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">column_order</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">delim</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'/'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#thoraxe.subexons.ases.get_transcript_scores" title="Permalink to this definition"></a></dt>
<dd><p>Return a DataFrame with the needed data to choose the canonical path.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="thoraxe.subexons.ases.read_splice_graph">
<span class="sig-prename descclassname"><span class="pre">thoraxe.subexons.ases.</span></span><span class="sig-name descname"><span class="pre">read_splice_graph</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">graph_file_name</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#thoraxe.subexons.ases.read_splice_graph" title="Permalink to this definition"></a></dt>
<dd><p>Read ThorAxe’s splice graph using NetworkX.</p>
</dd></dl>

</section>
<section id="module-thoraxe.subexons.graph">
<span id="thoraxe-subexons-graph-module"></span><h2>thoraxe.subexons.graph module<a class="headerlink" href="#module-thoraxe.subexons.graph" title="Permalink to this heading"></a></h2>
<p>graph: Functions to store the splice graph with conservation information.</p>
<dl class="py function">
<dt class="sig sig-object py" id="thoraxe.subexons.graph.nodes_and_edges2genes_and_transcripts">
<span class="sig-prename descclassname"><span class="pre">thoraxe.subexons.graph.</span></span><span class="sig-name descname"><span class="pre">nodes_and_edges2genes_and_transcripts</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#thoraxe.subexons.graph.nodes_and_edges2genes_and_transcripts" title="Permalink to this definition"></a></dt>
<dd><p>Return five dictionaries:</p>
<ul class="simple">
<li><p>nodes to genes</p></li>
<li><p>edges to genes</p></li>
<li><p>nodes to transcripts</p></li>
<li><p>edges to transcripts</p></li>
<li><p>edges to the transcript weighted conservation score</p></li>
</ul>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="thoraxe.subexons.graph.splice_graph_gml">
<span class="sig-prename descclassname"><span class="pre">thoraxe.subexons.graph.</span></span><span class="sig-name descname"><span class="pre">splice_graph_gml</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filename</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">s_exons_msas</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">node2genes</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">edge2genes</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">node2transcripts</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">edge2transcripts</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">edge2trx_cons</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">s_exon_2_char</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#thoraxe.subexons.graph.splice_graph_gml" title="Permalink to this definition"></a></dt>
<dd><p>Store the splice graph in GML (Graph Modelling Language) format.</p>
<p>It stores the conservation level information for nodes and edges, where
conservation is the fraction of species that has that particular feature.</p>
</dd></dl>

</section>
<section id="module-thoraxe.subexons.phylosofs">
<span id="thoraxe-subexons-phylosofs-module"></span><h2>thoraxe.subexons.phylosofs module<a class="headerlink" href="#module-thoraxe.subexons.phylosofs" title="Permalink to this heading"></a></h2>
<p>phylosofs: Functions to generate PhyloSofS input.</p>
<dl class="py function">
<dt class="sig sig-object py" id="thoraxe.subexons.phylosofs.get_s_exon2char">
<span class="sig-prename descclassname"><span class="pre">thoraxe.subexons.phylosofs.</span></span><span class="sig-name descname"><span class="pre">get_s_exon2char</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">s_exons</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#thoraxe.subexons.phylosofs.get_s_exon2char" title="Permalink to this definition"></a></dt>
<dd><p>Return a dictionary from orthologous regions (s-exon) to a single character.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">result</span> <span class="o">=</span> <span class="n">get_s_exon2char</span><span class="p">([</span><span class="s1">&#39;1_0&#39;</span><span class="p">,</span> <span class="s1">&#39;1_1&#39;</span><span class="p">,</span> <span class="s1">&#39;2_0&#39;</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">sorted</span><span class="p">(</span><span class="n">result</span><span class="p">)</span>
<span class="go">[&#39;1_0&#39;, &#39;1_1&#39;, &#39;2_0&#39;]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="p">[</span><span class="n">result</span><span class="p">[</span><span class="n">s_exon</span><span class="p">]</span> <span class="k">for</span> <span class="n">s_exon</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">result</span><span class="p">)]</span>
<span class="go">[&#39;%&#39;, &#39;(&#39;, &#39;)&#39;]</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="thoraxe.subexons.phylosofs.get_transcript2phylosofs">
<span class="sig-prename descclassname"><span class="pre">thoraxe.subexons.phylosofs.</span></span><span class="sig-name descname"><span class="pre">get_transcript2phylosofs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">s_exon2char</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#thoraxe.subexons.phylosofs.get_transcript2phylosofs" title="Permalink to this definition"></a></dt>
<dd><p>Return a dictionary from transcript id to its phylosofs representation.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="thoraxe.subexons.phylosofs.phylosofs_inputs">
<span class="sig-prename descclassname"><span class="pre">thoraxe.subexons.phylosofs.</span></span><span class="sig-name descname"><span class="pre">phylosofs_inputs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">s_exon_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ensembl_folder</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_folder</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#thoraxe.subexons.phylosofs.phylosofs_inputs" title="Permalink to this definition"></a></dt>
<dd><p>Create the input files needed for running PhyloSofS in a phylosofs folder.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="thoraxe.subexons.phylosofs.prune_tree">
<span class="sig-prename descclassname"><span class="pre">thoraxe.subexons.phylosofs.</span></span><span class="sig-name descname"><span class="pre">prune_tree</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input_tree</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_tree</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exontable_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">used_genes</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#thoraxe.subexons.phylosofs.prune_tree" title="Permalink to this definition"></a></dt>
<dd><p>Delete unused proteins from the Ensembl gene tree.</p>
<p>The function reads the Newick tree from input_tree and stores the pruned
tree in output_tree and returning the output_tree path. If an
error occurs, the function does not save the tree and raises an Exception.</p>
<p>In the tree, each protein is identified by their ensembl translation id.
For each gene, Ensembl uses the longest translated transcript sequence.</p>
<p>The exontable_file downloaded from Ensembl with transcript_query is needed
for getting the mapping between translation and gene ids of the used_genes.</p>
</dd></dl>

</section>
<section id="module-thoraxe.subexons.plot">
<span id="thoraxe-subexons-plot-module"></span><h2>thoraxe.subexons.plot module<a class="headerlink" href="#module-thoraxe.subexons.plot" title="Permalink to this heading"></a></h2>
<p>plot: Generate js data for plotting the MSA matrix using plotly.</p>
<dl class="py function">
<dt class="sig sig-object py" id="thoraxe.subexons.plot.create_python_structure">
<span class="sig-prename descclassname"><span class="pre">thoraxe.subexons.plot.</span></span><span class="sig-name descname"><span class="pre">create_python_structure</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cluster2data</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#thoraxe.subexons.plot.create_python_structure" title="Permalink to this definition"></a></dt>
<dd><p>Python structured data to be translated to JavaScript.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="thoraxe.subexons.plot.plot_msa_subexons">
<span class="sig-prename descclassname"><span class="pre">thoraxe.subexons.plot.</span></span><span class="sig-name descname"><span class="pre">plot_msa_subexons</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cluster2data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_folder</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#thoraxe.subexons.plot.plot_msa_subexons" title="Permalink to this definition"></a></dt>
<dd><p>Save html and js data in the output_folder.</p>
</dd></dl>

</section>
<section id="module-thoraxe.subexons.rescue">
<span id="thoraxe-subexons-rescue-module"></span><h2>thoraxe.subexons.rescue module<a class="headerlink" href="#module-thoraxe.subexons.rescue" title="Permalink to this heading"></a></h2>
<p>Functions for the rescue phase.</p>
<dl class="py function">
<dt class="sig sig-object py" id="thoraxe.subexons.rescue.get_unaligned_regions">
<span class="sig-prename descclassname"><span class="pre">thoraxe.subexons.rescue.</span></span><span class="sig-name descname"><span class="pre">get_unaligned_regions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq_i</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seq_j</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">minimum_len</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">4</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#thoraxe.subexons.rescue.get_unaligned_regions" title="Permalink to this definition"></a></dt>
<dd><p>Return the ungapped regions in <cite>seq_j</cite> that do not align with <cite>seq_i</cite>.</p>
<p>Returned regions should have at least <cite>minimum_len</cite>.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">seq_i</span> <span class="o">=</span> <span class="s1">&#39;------MHKCLVDE------YTEDQGGFRK------&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq_j</span> <span class="o">=</span> <span class="s1">&#39;MLLHYHHHKC-------LMCYTRDLHG---IH-L-K&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">get_unaligned_regions</span><span class="p">(</span><span class="n">seq_i</span><span class="p">,</span> <span class="n">seq_j</span><span class="p">)</span>
<span class="go">[&#39;MLLHYH&#39;, &#39;IHLK&#39;]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq_i</span> <span class="o">=</span> <span class="s1">&#39;XXXXXXXXXXMHKCLVDE------YTEDQGGFRK&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq_j</span> <span class="o">=</span> <span class="s1">&#39;----MLLHYHHHKC-------LMCYTRDLHG---&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">get_unaligned_regions</span><span class="p">(</span><span class="n">seq_i</span><span class="p">,</span> <span class="n">seq_j</span><span class="p">)</span>
<span class="go">[&#39;MLLHYH&#39;]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq_i</span> <span class="o">=</span> <span class="s1">&#39;XXXXXXXXXXMHKCLVDE------YTEDQGGFRK&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">seq_j</span> <span class="o">=</span> <span class="s1">&#39;----MLLHYHHHKC-----XXLMCYTRDLHG---&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">get_unaligned_regions</span><span class="p">(</span><span class="n">seq_i</span><span class="p">,</span> <span class="n">seq_j</span><span class="p">)</span>
<span class="go">[&#39;MLLHYH&#39;, &#39;XXLMC&#39;]</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="thoraxe.subexons.rescue.modify_subexon_cluster">
<span class="sig-prename descclassname"><span class="pre">thoraxe.subexons.rescue.</span></span><span class="sig-name descname"><span class="pre">modify_subexon_cluster</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">subexon_table</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subexon2cluster</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#thoraxe.subexons.rescue.modify_subexon_cluster" title="Permalink to this definition"></a></dt>
<dd><p>Modify subexon_table using the mapping in subexon2cluster.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="thoraxe.subexons.rescue.subexon_rescue_phase">
<span class="sig-prename descclassname"><span class="pre">thoraxe.subexons.rescue.</span></span><span class="sig-name descname"><span class="pre">subexon_rescue_phase</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cluster2data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subexon_table</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">minimum_len</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">4</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">coverage_cutoff</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">80.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">percent_identity_cutoff</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">30.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gap_open_penalty</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">-10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gap_extend_penalty</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">-1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">substitution_matrix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#thoraxe.subexons.rescue.subexon_rescue_phase" title="Permalink to this definition"></a></dt>
<dd><p>Execute the subexon rescue phase.</p>
<p>This function looks the subexons to rescues in the subexon_table by
looking at the ‘Cluster’ numbers that are lower than 0, i.e. -1 means that
the subexon has been cleaned up from ‘Cluster’ 1.
Then, each subexon to rescue is aligned against the subexons in other
‘Cluster’s, different from the original one. If it is nicely aligned
against another cluster, the ‘Cluster’ number is modified in the
subexon_table to match the new proposed ‘Cluster’ and the data relative to
the original pairwise alignment are cleaned up.
This function returns the list of ‘Cluster’s that have been modified.</p>
</dd></dl>

</section>
<section id="module-thoraxe.subexons.subexons">
<span id="thoraxe-subexons-subexons-module"></span><h2>thoraxe.subexons.subexons module<a class="headerlink" href="#module-thoraxe.subexons.subexons" title="Permalink to this heading"></a></h2>
<p>Functions to create the subexon table.</p>
<dl class="py class">
<dt class="sig sig-object py" id="thoraxe.subexons.subexons.Interval">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">thoraxe.subexons.subexons.</span></span><span class="sig-name descname"><span class="pre">Interval</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">start</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">end</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">components</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#thoraxe.subexons.subexons.Interval" title="Permalink to this definition"></a></dt>
<dd><dl class="py property">
<dt class="sig sig-object py" id="thoraxe.subexons.subexons.Interval.components">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">components</span></span><a class="headerlink" href="#thoraxe.subexons.subexons.Interval.components" title="Permalink to this definition"></a></dt>
<dd><p>Alias for field number 2</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="thoraxe.subexons.subexons.Interval.end">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">end</span></span><a class="headerlink" href="#thoraxe.subexons.subexons.Interval.end" title="Permalink to this definition"></a></dt>
<dd><p>Alias for field number 1</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="thoraxe.subexons.subexons.Interval.start">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">start</span></span><a class="headerlink" href="#thoraxe.subexons.subexons.Interval.start" title="Permalink to this definition"></a></dt>
<dd><p>Alias for field number 0</p>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="thoraxe.subexons.subexons.create_subexon_table">
<span class="sig-prename descclassname"><span class="pre">thoraxe.subexons.subexons.</span></span><span class="sig-name descname"><span class="pre">create_subexon_table</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">transcript_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">merge_non_redundant</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#thoraxe.subexons.subexons.create_subexon_table" title="Permalink to this definition"></a></dt>
<dd><p>Return a subexon table.</p>
<p>If merge_non_redundant is True (default), contiguous subexons that
appear together in only one exon are merged/joint. This happens
because subexons are defined using coordinates at the genomic level
to reduce redundancy.
However, some subexons does not have the same protein sequence because
of different reading frames/phases, giving non-redundant subexons at
the protein level.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="thoraxe.subexons.subexons.disjoint_intervals">
<span class="sig-prename descclassname"><span class="pre">thoraxe.subexons.subexons.</span></span><span class="sig-name descname"><span class="pre">disjoint_intervals</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">starts</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ends</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ids</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#thoraxe.subexons.subexons.disjoint_intervals" title="Permalink to this definition"></a></dt>
<dd><p>Return list of disjoint intervals.</p>
<p>Produce the set of disjoint intervals from a list of start and end
positions (same chr) and optionally ids.
Empty intervals are reported as well.
!!! The intervals need to be sorted by start and end position before
calling the function and the coordinates should be inclusive.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">disjoint_intervals</span><span class="p">([</span><span class="mi">1</span><span class="p">,</span><span class="mi">10</span><span class="p">,</span><span class="mi">20</span><span class="p">,</span><span class="mi">40</span><span class="p">],</span> <span class="p">[</span><span class="mi">9</span><span class="p">,</span><span class="mi">30</span><span class="p">,</span><span class="mi">30</span><span class="p">,</span><span class="mi">50</span><span class="p">],</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">])</span>
<span class="go">[Interval(start=1, end=9, components={0}), Interval(start=10, end=19, components={1}), Interval(start=20, end=30, components={1, 2}), Interval(start=31, end=39, components=set()), Interval(start=40, end=50, components={3})]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">disjoint_intervals</span><span class="p">([</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">],</span> <span class="p">[</span><span class="mi">9</span><span class="p">,</span><span class="mi">7</span><span class="p">],</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">])</span>
<span class="go">[Interval(start=1, end=7, components={0, 1}), Interval(start=8, end=9, components={0})]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">disjoint_intervals</span><span class="p">([</span><span class="mi">1</span><span class="p">,</span><span class="mi">3</span><span class="p">],</span> <span class="p">[</span><span class="mi">7</span><span class="p">,</span><span class="mi">9</span><span class="p">],</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">])</span>
<span class="go">[Interval(start=1, end=2, components={0}), Interval(start=3, end=7, components={0, 1}), Interval(start=8, end=9, components={1})]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">disjoint_intervals</span><span class="p">([</span><span class="mi">3</span><span class="p">,</span><span class="mi">1</span><span class="p">],</span> <span class="p">[</span><span class="mi">9</span><span class="p">,</span><span class="mi">7</span><span class="p">],</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">])</span>
<span class="gt">Traceback (most recent call last):</span>
<span class="w">    </span><span class="o">...</span>
<span class="gr">AssertionError</span>: <span class="n">starts should be sorted.</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">disjoint_intervals</span><span class="p">([</span><span class="mi">3</span><span class="p">,</span><span class="mi">1</span><span class="p">],</span> <span class="p">[</span><span class="mi">9</span><span class="p">,</span><span class="mi">7</span><span class="p">,</span><span class="mi">10</span><span class="p">],</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">])</span>
<span class="gt">Traceback (most recent call last):</span>
<span class="w">    </span><span class="o">...</span>
<span class="gr">AssertionError</span>: <span class="n">starts, ends and ids should have the same length.</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="thoraxe.subexons.subexons.update_to_merge_list">
<span class="sig-prename descclassname"><span class="pre">thoraxe.subexons.subexons.</span></span><span class="sig-name descname"><span class="pre">update_to_merge_list</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">to_merge</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subexon_1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subexon_2</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#thoraxe.subexons.subexons.update_to_merge_list" title="Permalink to this definition"></a></dt>
<dd><p>Add subexon_1 and subexon_2 to the to_merge list.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">update_to_merge_list</span><span class="p">([],</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span>
<span class="go">[{1, 2}]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">update_to_merge_list</span><span class="p">([{</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">}],</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">)</span>
<span class="go">[{1, 2, 3}]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">update_to_merge_list</span><span class="p">([{</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">}],</span> <span class="mi">8</span><span class="p">,</span> <span class="mi">9</span><span class="p">)</span>
<span class="go">[{1, 2}, {8, 9}]</span>
</pre></div>
</div>
</dd></dl>

</section>
<section id="module-thoraxe.subexons.tidy">
<span id="thoraxe-subexons-tidy-module"></span><h2>thoraxe.subexons.tidy module<a class="headerlink" href="#module-thoraxe.subexons.tidy" title="Permalink to this heading"></a></h2>
<p>tidy:  Utils for having a tidy CSV table output (denormalized table).</p>
<dl class="py function">
<dt class="sig sig-object py" id="thoraxe.subexons.tidy.get_tidy_table">
<span class="sig-prename descclassname"><span class="pre">thoraxe.subexons.tidy.</span></span><span class="sig-name descname"><span class="pre">get_tidy_table</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">table</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene2species</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#thoraxe.subexons.tidy.get_tidy_table" title="Permalink to this definition"></a></dt>
<dd><p>Save a tidy and denormalized output table.</p>
<p>Takes the final exon table and a dict from gene to species name to save a
csv file with the output table in the outputdir.</p>
</dd></dl>

</section>
<section id="module-thoraxe.subexons">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-thoraxe.subexons" title="Permalink to this heading"></a></h2>
<p>subexons: Module to identify and cut subexons.</p>
<p>This defines functions that take as input the DataFrame output from
transcript_info.read_transcript_info.</p>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="thoraxe.add_transcripts.html" class="btn btn-neutral float-left" title="thoraxe.add_transcripts package" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="thoraxe.transcript_info.html" class="btn btn-neutral float-right" title="thoraxe.transcript_info package" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2019, Diego Javier Zea, Hugues Richard &amp; Elodie Laine.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>